// @amplitude/utils/dist/src/index.js
Object.defineProperty(exports, "__esModule", { value: true });
var base36_1 = require("./base36");
Object.defineProperty(exports, "generateBase36Id", { enumerable: true, get: function () { return base36_1.generateBase36Id; } });
Object.defineProperty(exports, "DEVICE_ID_LENGTH", { enumerable: true, get: function () { return base36_1.DEVICE_ID_LENGTH; } });
var logger_1 = require("./logger");
Object.defineProperty(exports, "logger", { enumerable: true, get: function () { return logger_1.logger; } });
var misc_1 = require("./misc");
Object.defineProperty(exports, "asyncSleep", { enumerable: true, get: function () { return misc_1.asyncSleep; } });
Object.defineProperty(exports, "getGlobalAmplitudeNamespace", { enumerable: true, get: function () { return misc_1.getGlobalAmplitudeNamespace; } });
Object.defineProperty(exports, "isBrowserEnv", { enumerable: true, get: function () { return misc_1.isBrowserEnv; } });
Object.defineProperty(exports, "isNodeEnv", { enumerable: true, get: function () { return misc_1.isNodeEnv; } });
Object.defineProperty(exports, "prototypeJsFix", { enumerable: true, get: function () { return misc_1.prototypeJsFix; } });
var queue_1 = require("./queue");
Object.defineProperty(exports, "AsyncQueue", { enumerable: true, get: function () { return queue_1.AsyncQueue; } });
var response_1 = require("./response");
Object.defineProperty(exports, "collectInvalidEventIndices", { enumerable: true, get: function () { return response_1.collectInvalidEventIndices; } });
Object.defineProperty(exports, "mapHttpMessageToResponse", { enumerable: true, get: function () { return response_1.mapHttpMessageToResponse; } });
Object.defineProperty(exports, "mapJSONToResponse", { enumerable: true, get: function () { return response_1.mapJSONToResponse; } });
var status_1 = require("./status");
Object.defineProperty(exports, "mapHttpCodeToStatus", { enumerable: true, get: function () { return status_1.mapHttpCodeToStatus; } });
var validate_1 = require("./validate");
Object.defineProperty(exports, "isValidEvent", { enumerable: true, get: function () { return validate_1.isValidEvent; } });
var validateProperties_1 = require("./validateProperties");
Object.defineProperty(exports, "isValidProperties", { enumerable: true, get: function () { return validateProperties_1.isValidProperties; } });
