// @terra-money/legacy.proto/cosmos/mint/v1beta1/mint.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = exports.Minter = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "cosmos.mint.v1beta1";
const baseMinter = { inflation: "", annualProvisions: "" };
exports.Minter = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.inflation !== "") {
            writer.uint32(10).string(message.inflation);
        }
        if (message.annualProvisions !== "") {
            writer.uint32(18).string(message.annualProvisions);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = Object.assign({}, baseMinter);
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.inflation = reader.string();
                    break;
                case 2:
                    message.annualProvisions = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const message = Object.assign({}, baseMinter);
        if (object.inflation !== undefined && object.inflation !== null) {
            message.inflation = String(object.inflation);
        }
        else {
            message.inflation = "";
        }
        if (object.annualProvisions !== undefined && object.annualProvisions !== null) {
            message.annualProvisions = String(object.annualProvisions);
        }
        else {
            message.annualProvisions = "";
        }
        return message;
    },
    toJSON(message) {
        const obj = {};
        message.inflation !== undefined && (obj.inflation = message.inflation);
        message.annualProvisions !== undefined && (obj.annualProvisions = message.annualProvisions);
        return obj;
    },
    fromPartial(object) {
        const message = Object.assign({}, baseMinter);
        if (object.inflation !== undefined && object.inflation !== null) {
            message.inflation = object.inflation;
        }
        else {
            message.inflation = "";
        }
        if (object.annualProvisions !== undefined && object.annualProvisions !== null) {
            message.annualProvisions = object.annualProvisions;
        }
        else {
            message.annualProvisions = "";
        }
        return message;
    },
};
const baseParams = {
    mintDenom: "",
    inflationRateChange: "",
    inflationMax: "",
    inflationMin: "",
    goalBonded: "",
    blocksPerYear: long_1.default.UZERO,
};
exports.Params = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.mintDenom !== "") {
            writer.uint32(10).string(message.mintDenom);
        }
        if (message.inflationRateChange !== "") {
            writer.uint32(18).string(message.inflationRateChange);
        }
        if (message.inflationMax !== "") {
            writer.uint32(26).string(message.inflationMax);
        }
        if (message.inflationMin !== "") {
            writer.uint32(34).string(message.inflationMin);
        }
        if (message.goalBonded !== "") {
            writer.uint32(42).string(message.goalBonded);
        }
        if (!message.blocksPerYear.isZero()) {
            writer.uint32(48).uint64(message.blocksPerYear);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = Object.assign({}, baseParams);
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.mintDenom = reader.string();
                    break;
                case 2:
                    message.inflationRateChange = reader.string();
                    break;
                case 3:
                    message.inflationMax = reader.string();
                    break;
                case 4:
                    message.inflationMin = reader.string();
                    break;
                case 5:
                    message.goalBonded = reader.string();
                    break;
                case 6:
                    message.blocksPerYear = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const message = Object.assign({}, baseParams);
        if (object.mintDenom !== undefined && object.mintDenom !== null) {
            message.mintDenom = String(object.mintDenom);
        }
        else {
            message.mintDenom = "";
        }
        if (object.inflationRateChange !== undefined && object.inflationRateChange !== null) {
            message.inflationRateChange = String(object.inflationRateChange);
        }
        else {
            message.inflationRateChange = "";
        }
        if (object.inflationMax !== undefined && object.inflationMax !== null) {
            message.inflationMax = String(object.inflationMax);
        }
        else {
            message.inflationMax = "";
        }
        if (object.inflationMin !== undefined && object.inflationMin !== null) {
            message.inflationMin = String(object.inflationMin);
        }
        else {
            message.inflationMin = "";
        }
        if (object.goalBonded !== undefined && object.goalBonded !== null) {
            message.goalBonded = String(object.goalBonded);
        }
        else {
            message.goalBonded = "";
        }
        if (object.blocksPerYear !== undefined && object.blocksPerYear !== null) {
            message.blocksPerYear = long_1.default.fromString(object.blocksPerYear);
        }
        else {
            message.blocksPerYear = long_1.default.UZERO;
        }
        return message;
    },
    toJSON(message) {
        const obj = {};
        message.mintDenom !== undefined && (obj.mintDenom = message.mintDenom);
        message.inflationRateChange !== undefined && (obj.inflationRateChange = message.inflationRateChange);
        message.inflationMax !== undefined && (obj.inflationMax = message.inflationMax);
        message.inflationMin !== undefined && (obj.inflationMin = message.inflationMin);
        message.goalBonded !== undefined && (obj.goalBonded = message.goalBonded);
        message.blocksPerYear !== undefined &&
            (obj.blocksPerYear = (message.blocksPerYear || long_1.default.UZERO).toString());
        return obj;
    },
    fromPartial(object) {
        const message = Object.assign({}, baseParams);
        if (object.mintDenom !== undefined && object.mintDenom !== null) {
            message.mintDenom = object.mintDenom;
        }
        else {
            message.mintDenom = "";
        }
        if (object.inflationRateChange !== undefined && object.inflationRateChange !== null) {
            message.inflationRateChange = object.inflationRateChange;
        }
        else {
            message.inflationRateChange = "";
        }
        if (object.inflationMax !== undefined && object.inflationMax !== null) {
            message.inflationMax = object.inflationMax;
        }
        else {
            message.inflationMax = "";
        }
        if (object.inflationMin !== undefined && object.inflationMin !== null) {
            message.inflationMin = object.inflationMin;
        }
        else {
            message.inflationMin = "";
        }
        if (object.goalBonded !== undefined && object.goalBonded !== null) {
            message.goalBonded = object.goalBonded;
        }
        else {
            message.goalBonded = "";
        }
        if (object.blocksPerYear !== undefined && object.blocksPerYear !== null) {
            message.blocksPerYear = object.blocksPerYear;
        }
        else {
            message.blocksPerYear = long_1.default.UZERO;
        }
        return message;
    },
};
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
