// @liquality/cryptoassets/dist/src/chains/testnet/evm/optimism.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const optimism_1 = tslib_1.__importDefault(require("../../../chains/mainnet/evm/optimism"));
const utils_1 = require("../../../chains/utils");
const optimismTestnet = (0, utils_1.transformMainnetToTestnetChain)(optimism_1.default, {
    name: 'optimism_testnet',
    coinType: '60',
    networkId: 69,
    chainId: 69,
    isTestnet: true,
    rpcUrls: ['https://optimism-kovan.infura.io/v3/da99ebc8c0964bb8bb757b6f8cc40f1f'],
}, [
    {
        tx: 'https://kovan-optimistic.etherscan.io/tx/{hash}',
        address: 'https://kovan-optimistic.etherscan.io/address/{address}',
        token: 'https://kovan-optimistic.etherscan.io/token/{token}',
    },
], 'https://optimismfaucet.xyz/');
exports.default = Object.assign(Object.assign({}, optimismTestnet), { gasLimit: {
        send: {
            native: 21000,
            nonNative: 100000,
        },
        sendL1: {
            native: 7500,
            nonNative: 8250,
        },
    } });
