// @chainify/evm/dist/lib/typechain/factories/ILiqualityHTLC__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ILiqualityHTLC__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "LiqualityHTLC__InvalidExpiration",
        type: "error",
    },
    {
        inputs: [],
        name: "LiqualityHTLC__InvalidMsgValue",
        type: "error",
    },
    {
        inputs: [],
        name: "LiqualityHTLC__InvalidSender",
        type: "error",
    },
    {
        inputs: [],
        name: "LiqualityHTLC__InvalidSwapAmount",
        type: "error",
    },
    {
        inputs: [],
        name: "LiqualityHTLC__SwapAlreadyExists",
        type: "error",
    },
    {
        inputs: [],
        name: "LiqualityHTLC__SwapDoesNotExist",
        type: "error",
    },
    {
        inputs: [],
        name: "LiqualityHTLC__SwapNotExpired",
        type: "error",
    },
    {
        inputs: [],
        name: "LiqualityHTLC__WrongSecret",
        type: "error",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "bytes32",
                name: "id",
                type: "bytes32",
            },
            {
                indexed: false,
                internalType: "bytes32",
                name: "secret",
                type: "bytes32",
            },
        ],
        name: "Claim",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "bytes32",
                name: "id",
                type: "bytes32",
            },
            {
                components: [
                    {
                        internalType: "uint256",
                        name: "amount",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "expiration",
                        type: "uint256",
                    },
                    {
                        internalType: "bytes32",
                        name: "secretHash",
                        type: "bytes32",
                    },
                    {
                        internalType: "address",
                        name: "tokenAddress",
                        type: "address",
                    },
                    {
                        internalType: "address",
                        name: "refundAddress",
                        type: "address",
                    },
                    {
                        internalType: "address",
                        name: "recipientAddress",
                        type: "address",
                    },
                ],
                indexed: false,
                internalType: "struct ILiqualityHTLC.HTLCData",
                name: "htlc",
                type: "tuple",
            },
        ],
        name: "Initiate",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "bytes32",
                name: "id",
                type: "bytes32",
            },
        ],
        name: "Refund",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "id",
                type: "bytes32",
            },
            {
                internalType: "bytes32",
                name: "secret",
                type: "bytes32",
            },
        ],
        name: "claim",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                components: [
                    {
                        internalType: "uint256",
                        name: "amount",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "expiration",
                        type: "uint256",
                    },
                    {
                        internalType: "bytes32",
                        name: "secretHash",
                        type: "bytes32",
                    },
                    {
                        internalType: "address",
                        name: "tokenAddress",
                        type: "address",
                    },
                    {
                        internalType: "address",
                        name: "refundAddress",
                        type: "address",
                    },
                    {
                        internalType: "address",
                        name: "recipientAddress",
                        type: "address",
                    },
                ],
                internalType: "struct ILiqualityHTLC.HTLCData",
                name: "htlc",
                type: "tuple",
            },
        ],
        name: "initiate",
        outputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        stateMutability: "payable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "id",
                type: "bytes32",
            },
        ],
        name: "refund",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class ILiqualityHTLC__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.ILiqualityHTLC__factory = ILiqualityHTLC__factory;
ILiqualityHTLC__factory.abi = _abi;
