// @chainify/bitcoin/dist/lib/types.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SwapMode = exports.AddressType = exports.BitcoinJsonRpcTypes = exports.BitcoinEsploraTypes = void 0;
exports.BitcoinEsploraTypes = __importStar(require("./chain/esplora/types"));
exports.BitcoinJsonRpcTypes = __importStar(require("./chain/jsonRpc/types"));
__exportStar(require("./swap/types"), exports);
var AddressType;
(function (AddressType) {
    AddressType["LEGACY"] = "legacy";
    AddressType["P2SH_SEGWIT"] = "p2sh-segwit";
    AddressType["BECH32"] = "bech32";
})(AddressType = exports.AddressType || (exports.AddressType = {}));
var SwapMode;
(function (SwapMode) {
    SwapMode["P2SH"] = "p2sh";
    SwapMode["P2SH_SEGWIT"] = "p2shSegwit";
    SwapMode["P2WSH"] = "p2wsh";
})(SwapMode = exports.SwapMode || (exports.SwapMode = {}));
