// @sinatdt/contracts-helper/dist/contracts/cc-exchange-router-contract-interface.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
const EthereumBase = require("../ethereum-base");
const { CCExchangeRouterABI } = require("@sinatdt/configs").teleswap.ABI;
class CcExchangeRouter extends EthereumBase {
    constructor(connectionInfo, contractAddress) {
        super(connectionInfo);
        this.contractAddress = contractAddress;
        this.contract = new this.web3Eth.Contract(CCExchangeRouterABI, contractAddress);
    }
    getProtocolPercentageFee() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.contract.methods.protocolPercentageFee().call();
        });
    }
    isUsed(txId) {
        return __awaiter(this, void 0, void 0, function* () {
            let txIdWith0x = `0x${Buffer.from(txId, "hex").reverse().toString("hex")}`;
            let isUsed = yield this.contract.methods.isRequestUsed(txIdWith0x).call();
            return isUsed;
        });
    }
    sendExchangeRequest(lockerScript, parsedTx, merkleProof, blockNumber, blockFee, nonce = undefined) {
        return __awaiter(this, void 0, void 0, function* () {
            let value = blockFee ? (+blockFee * 1.3).toFixed() : "100000000000000000";
            const gasAmount = yield this.contract.methods
                .ccExchange(parsedTx.version, parsedTx.vin, parsedTx.vout, parsedTx.locktime, blockNumber, merkleProof.intermediateNodes, merkleProof.transactionIndex, `0x${lockerScript}`)
                .estimateGas({ from: this.currentAccount, value });
            let response = yield this.contract.methods
                .ccExchange(parsedTx.version, parsedTx.vin, parsedTx.vout, parsedTx.locktime, blockNumber, merkleProof.intermediateNodes, merkleProof.transactionIndex, `0x${lockerScript}`)
                .send({ from: this.currentAccount, gas: gasAmount, value, nonce })
                .then((recipient) => ({
                success: recipient.status,
                txId: recipient.transactionHash,
            }));
            return response;
        });
    }
}
module.exports = CcExchangeRouter;
