// @sinatdt/contracts-helper/dist/contracts/lending-router-contract-interface.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
const EthereumBase = require("../ethereum-base");
const { LendingRouterABI } = require("@sinatdt/configs").teleswap.ABI;
class LendingRouter extends EthereumBase {
    constructor(connectionInfo, contractAddress) {
        super(connectionInfo);
        this.contractAddress = contractAddress;
        this.contract = new this.web3Eth.Contract(LendingRouterABI, contractAddress);
        this.contract.handleRevert = true;
    }
    isUsed(txId) {
        return __awaiter(this, void 0, void 0, function* () {
            let txIdWith0x = `0x${Buffer.from(txId, "hex").reverse().toString("hex")}`;
            let isUsed = yield this.contract.methods
                .CCLendingRequests(txIdWith0x)
                .call();
            return isUsed === null || isUsed === void 0 ? void 0 : isUsed.isUsed;
        });
    }
    sendLendingRequest(lockerScript, parsedTx, merkleProof, blockNumber, blockFee, nonce = undefined) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                let value = blockFee ? (+blockFee * 1.3).toFixed() : "100000000000000000";
                const gasAmount = yield this.contract.methods
                    .lending(parsedTx.version, parsedTx.vin, parsedTx.vout, parsedTx.locktime, blockNumber, merkleProof.intermediateNodes, merkleProof.transactionIndex, `0x${lockerScript}`)
                    .estimateGas({ from: this.currentAccount, value, nonce });
                let response = yield this.contract.methods
                    .lending(parsedTx.version, parsedTx.vin, parsedTx.vout, parsedTx.locktime, blockNumber, merkleProof.intermediateNodes, merkleProof.transactionIndex, `0x${lockerScript}`)
                    .send({ from: this.currentAccount, gas: gasAmount, value, nonce })
                    .then((recipient) => ({
                    success: recipient.status,
                    txId: recipient.transactionHash,
                }));
                return response;
            }
            catch (error) {
                return {
                    success: false,
                    message: error.message,
                };
            }
        });
    }
}
module.exports = LendingRouter;
