// @sinatdt/contracts-helper/dist/contracts/wrapped-token-contract-interface.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
const EthereumBase = require("../ethereum-base");
const { WrappedTokenABI } = require("@sinatdt/configs").teleswap.ABI;
class WrappedToken extends EthereumBase {
    constructor(connectionConfig, contractAddress, unit) {
        super(connectionConfig);
        this.contractAddress = contractAddress;
        this.contract = new this.web3Eth.Contract(WrappedTokenABI, contractAddress);
        this.unit = unit;
    }
    setDecimal() {
        return __awaiter(this, void 0, void 0, function* () {
            let decimal = yield this.contract.methods.decimals().call();
            this.unit = decimal;
        });
    }
    getDecimal() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.unit)
                yield this.setDecimal();
            return this.unit;
        });
    }
    mintTestToken() {
        return __awaiter(this, void 0, void 0, function* () {
            const gasAmount = yield this.contract.methods
                .mintTestToken()
                .estimateGas({ from: this.currentAccount });
            return this.contract.methods
                .mintTestToken()
                .send({ from: this.currentAccount, gas: gasAmount });
        });
    }
    approve(address, amount) {
        return __awaiter(this, void 0, void 0, function* () {
            const gasAmount = yield this.contract.methods
                .approve(address, amount)
                .estimateGas({ from: this.currentAccount });
            return this.contract.methods
                .approve(address, amount)
                .send({ from: this.currentAccount, gas: gasAmount });
        });
    }
    getBalance(address) {
        return __awaiter(this, void 0, void 0, function* () {
            let balance = yield this.contract.methods
                .balanceOf(address || this.currentAccount)
                .call();
            return balance;
        });
    }
    getApprovedBalanceForAddress(address) {
        return __awaiter(this, void 0, void 0, function* () {
            let balance = yield this.contract.methods
                .allowance(this.currentAccount, address)
                .call();
            return balance;
        });
    }
}
module.exports = WrappedToken;
