// @terra-money/terra.proto/ibc/applications/fee/v1/ack.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IncentivizedAcknowledgement = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "ibc.applications.fee.v1";
const baseIncentivizedAcknowledgement = { forwardRelayerAddress: "", underlyingAppSuccess: false };
exports.IncentivizedAcknowledgement = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.result.length !== 0) {
            writer.uint32(10).bytes(message.result);
        }
        if (message.forwardRelayerAddress !== "") {
            writer.uint32(18).string(message.forwardRelayerAddress);
        }
        if (message.underlyingAppSuccess === true) {
            writer.uint32(24).bool(message.underlyingAppSuccess);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = Object.assign({}, baseIncentivizedAcknowledgement);
        message.result = new Uint8Array();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.result = reader.bytes();
                    break;
                case 2:
                    message.forwardRelayerAddress = reader.string();
                    break;
                case 3:
                    message.underlyingAppSuccess = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const message = Object.assign({}, baseIncentivizedAcknowledgement);
        message.result = new Uint8Array();
        if (object.result !== undefined && object.result !== null) {
            message.result = bytesFromBase64(object.result);
        }
        if (object.forwardRelayerAddress !== undefined && object.forwardRelayerAddress !== null) {
            message.forwardRelayerAddress = String(object.forwardRelayerAddress);
        }
        else {
            message.forwardRelayerAddress = "";
        }
        if (object.underlyingAppSuccess !== undefined && object.underlyingAppSuccess !== null) {
            message.underlyingAppSuccess = Boolean(object.underlyingAppSuccess);
        }
        else {
            message.underlyingAppSuccess = false;
        }
        return message;
    },
    toJSON(message) {
        const obj = {};
        message.result !== undefined &&
            (obj.result = base64FromBytes(message.result !== undefined ? message.result : new Uint8Array()));
        message.forwardRelayerAddress !== undefined &&
            (obj.forwardRelayerAddress = message.forwardRelayerAddress);
        message.underlyingAppSuccess !== undefined && (obj.underlyingAppSuccess = message.underlyingAppSuccess);
        return obj;
    },
    fromPartial(object) {
        const message = Object.assign({}, baseIncentivizedAcknowledgement);
        if (object.result !== undefined && object.result !== null) {
            message.result = object.result;
        }
        else {
            message.result = new Uint8Array();
        }
        if (object.forwardRelayerAddress !== undefined && object.forwardRelayerAddress !== null) {
            message.forwardRelayerAddress = object.forwardRelayerAddress;
        }
        else {
            message.forwardRelayerAddress = "";
        }
        if (object.underlyingAppSuccess !== undefined && object.underlyingAppSuccess !== null) {
            message.underlyingAppSuccess = object.underlyingAppSuccess;
        }
        else {
            message.underlyingAppSuccess = false;
        }
        return message;
    },
};
var globalThis = (() => {
    if (typeof globalThis !== "undefined")
        return globalThis;
    if (typeof self !== "undefined")
        return self;
    if (typeof window !== "undefined")
        return window;
    if (typeof global !== "undefined")
        return global;
    throw "Unable to locate global object";
})();
const atob = globalThis.atob || ((b64) => globalThis.Buffer.from(b64, "base64").toString("binary"));
function bytesFromBase64(b64) {
    const bin = atob(b64);
    const arr = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; ++i) {
        arr[i] = bin.charCodeAt(i);
    }
    return arr;
}
const btoa = globalThis.btoa || ((bin) => globalThis.Buffer.from(bin, "binary").toString("base64"));
function base64FromBytes(arr) {
    const bin = [];
    for (const byte of arr) {
        bin.push(String.fromCharCode(byte));
    }
    return btoa(bin.join(""));
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
