// @chainify/near/dist/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NearWalletProvider = exports.NearUtils = exports.NearTypes = exports.NearSwapProvider = exports.NearNetworks = exports.NearChainProvider = void 0;
var NearChainProvider_1 = require("./chain/NearChainProvider");
Object.defineProperty(exports, "NearChainProvider", { enumerable: true, get: function () { return NearChainProvider_1.NearChainProvider; } });
var networks_1 = require("./networks");
Object.defineProperty(exports, "NearNetworks", { enumerable: true, get: function () { return networks_1.NearNetworks; } });
var NearSwapProvider_1 = require("./swap/NearSwapProvider");
Object.defineProperty(exports, "NearSwapProvider", { enumerable: true, get: function () { return NearSwapProvider_1.NearSwapProvider; } });
exports.NearTypes = __importStar(require("./types"));
exports.NearUtils = __importStar(require("./utils"));
var NearWalletProvider_1 = require("./wallet/NearWalletProvider");
Object.defineProperty(exports, "NearWalletProvider", { enumerable: true, get: function () { return NearWalletProvider_1.NearWalletProvider; } });
