// @chainify/near/dist/lib/types.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NearAccount = exports.BN = exports.Action = exports.parseSeedPhrase = exports.transactions = exports.providers = exports.keyStores = exports.KeyPair = exports.InMemorySigner = exports.Account = void 0;
const bn_js_1 = __importDefault(require("bn.js"));
exports.BN = bn_js_1.default;
const near_api_js_1 = require("near-api-js");
const transaction_1 = require("near-api-js/lib/transaction");
Object.defineProperty(exports, "Action", { enumerable: true, get: function () { return transaction_1.Action; } });
var near_api_js_2 = require("near-api-js");
Object.defineProperty(exports, "Account", { enumerable: true, get: function () { return near_api_js_2.Account; } });
Object.defineProperty(exports, "InMemorySigner", { enumerable: true, get: function () { return near_api_js_2.InMemorySigner; } });
Object.defineProperty(exports, "KeyPair", { enumerable: true, get: function () { return near_api_js_2.KeyPair; } });
Object.defineProperty(exports, "keyStores", { enumerable: true, get: function () { return near_api_js_2.keyStores; } });
Object.defineProperty(exports, "providers", { enumerable: true, get: function () { return near_api_js_2.providers; } });
Object.defineProperty(exports, "transactions", { enumerable: true, get: function () { return near_api_js_2.transactions; } });
var near_seed_phrase_1 = require("near-seed-phrase");
Object.defineProperty(exports, "parseSeedPhrase", { enumerable: true, get: function () { return near_seed_phrase_1.parseSeedPhrase; } });
class NearAccount extends near_api_js_1.Account {
    constructor(connection, accountId) {
        super(connection, accountId);
    }
    signAndSendTransaction({ receiverId, actions }) {
        const _super = Object.create(null, {
            signAndSendTransaction: { get: () => super.signAndSendTransaction }
        });
        return __awaiter(this, void 0, void 0, function* () {
            return _super.signAndSendTransaction.call(this, { receiverId, actions });
        });
    }
}
exports.NearAccount = NearAccount;
