// @unstoppabledomains/resolution/build/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProviderFactories = exports.Eip1193Factories = exports.Eip1993Factories = exports.DnsRecordsErrorCode = exports.DnsRecordsError = exports.ConfigurationErrorCode = exports.ConfigurationError = exports.ResolutionErrorCode = exports.ResolutionError = exports.DnsUtils = exports.default = exports.Resolution = void 0;
var Resolution_1 = require("./Resolution");
Object.defineProperty(exports, "Resolution", { enumerable: true, get: function () { return Resolution_1.Resolution; } });
Object.defineProperty(exports, "default", { enumerable: true, get: function () { return Resolution_1.Resolution; } });
__exportStar(require("./types/publicTypes"), exports);
var DnsUtils_1 = require("./utils/DnsUtils");
Object.defineProperty(exports, "DnsUtils", { enumerable: true, get: function () { return __importDefault(DnsUtils_1).default; } });
var resolutionError_1 = require("./errors/resolutionError");
Object.defineProperty(exports, "ResolutionError", { enumerable: true, get: function () { return resolutionError_1.ResolutionError; } });
Object.defineProperty(exports, "ResolutionErrorCode", { enumerable: true, get: function () { return resolutionError_1.ResolutionErrorCode; } });
var configurationError_1 = require("./errors/configurationError");
Object.defineProperty(exports, "ConfigurationError", { enumerable: true, get: function () { return configurationError_1.ConfigurationError; } });
Object.defineProperty(exports, "ConfigurationErrorCode", { enumerable: true, get: function () { return configurationError_1.ConfigurationErrorCode; } });
var dnsRecordsError_1 = require("./errors/dnsRecordsError");
Object.defineProperty(exports, "DnsRecordsError", { enumerable: true, get: function () { return dnsRecordsError_1.DnsRecordsError; } });
Object.defineProperty(exports, "DnsRecordsErrorCode", { enumerable: true, get: function () { return dnsRecordsError_1.DnsRecordsErrorCode; } });
var Eip1993Factories_1 = require("./utils/Eip1993Factories");
Object.defineProperty(exports, "Eip1993Factories", { enumerable: true, get: function () { return Eip1993Factories_1.Eip1993Factories; } });
var Eip1993Factories_2 = require("./utils/Eip1993Factories");
Object.defineProperty(exports, "Eip1193Factories", { enumerable: true, get: function () { return Eip1993Factories_2.Eip1993Factories; } });
var Eip1993Factories_3 = require("./utils/Eip1993Factories");
Object.defineProperty(exports, "ProviderFactories", { enumerable: true, get: function () { return Eip1993Factories_3.Eip1993Factories; } });
