// @terra-money/legacy.proto/terra/wasm/v1beta1/genesis.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Contract = exports.Code = exports.Model = exports.GenesisState = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const wasm_1 = require("../../../terra/wasm/v1beta1/wasm");
exports.protobufPackage = "terra.wasm.v1beta1";
const baseGenesisState = { lastCodeId: long_1.default.UZERO, lastInstanceId: long_1.default.UZERO };
exports.GenesisState = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.params !== undefined) {
            wasm_1.Params.encode(message.params, writer.uint32(10).fork()).ldelim();
        }
        if (!message.lastCodeId.isZero()) {
            writer.uint32(16).uint64(message.lastCodeId);
        }
        if (!message.lastInstanceId.isZero()) {
            writer.uint32(24).uint64(message.lastInstanceId);
        }
        for (const v of message.codes) {
            exports.Code.encode(v, writer.uint32(34).fork()).ldelim();
        }
        for (const v of message.contracts) {
            exports.Contract.encode(v, writer.uint32(42).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = Object.assign({}, baseGenesisState);
        message.codes = [];
        message.contracts = [];
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.params = wasm_1.Params.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.lastCodeId = reader.uint64();
                    break;
                case 3:
                    message.lastInstanceId = reader.uint64();
                    break;
                case 4:
                    message.codes.push(exports.Code.decode(reader, reader.uint32()));
                    break;
                case 5:
                    message.contracts.push(exports.Contract.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const message = Object.assign({}, baseGenesisState);
        message.codes = [];
        message.contracts = [];
        if (object.params !== undefined && object.params !== null) {
            message.params = wasm_1.Params.fromJSON(object.params);
        }
        else {
            message.params = undefined;
        }
        if (object.lastCodeId !== undefined && object.lastCodeId !== null) {
            message.lastCodeId = long_1.default.fromString(object.lastCodeId);
        }
        else {
            message.lastCodeId = long_1.default.UZERO;
        }
        if (object.lastInstanceId !== undefined && object.lastInstanceId !== null) {
            message.lastInstanceId = long_1.default.fromString(object.lastInstanceId);
        }
        else {
            message.lastInstanceId = long_1.default.UZERO;
        }
        if (object.codes !== undefined && object.codes !== null) {
            for (const e of object.codes) {
                message.codes.push(exports.Code.fromJSON(e));
            }
        }
        if (object.contracts !== undefined && object.contracts !== null) {
            for (const e of object.contracts) {
                message.contracts.push(exports.Contract.fromJSON(e));
            }
        }
        return message;
    },
    toJSON(message) {
        const obj = {};
        message.params !== undefined && (obj.params = message.params ? wasm_1.Params.toJSON(message.params) : undefined);
        message.lastCodeId !== undefined && (obj.lastCodeId = (message.lastCodeId || long_1.default.UZERO).toString());
        message.lastInstanceId !== undefined &&
            (obj.lastInstanceId = (message.lastInstanceId || long_1.default.UZERO).toString());
        if (message.codes) {
            obj.codes = message.codes.map((e) => (e ? exports.Code.toJSON(e) : undefined));
        }
        else {
            obj.codes = [];
        }
        if (message.contracts) {
            obj.contracts = message.contracts.map((e) => (e ? exports.Contract.toJSON(e) : undefined));
        }
        else {
            obj.contracts = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = Object.assign({}, baseGenesisState);
        message.codes = [];
        message.contracts = [];
        if (object.params !== undefined && object.params !== null) {
            message.params = wasm_1.Params.fromPartial(object.params);
        }
        else {
            message.params = undefined;
        }
        if (object.lastCodeId !== undefined && object.lastCodeId !== null) {
            message.lastCodeId = object.lastCodeId;
        }
        else {
            message.lastCodeId = long_1.default.UZERO;
        }
        if (object.lastInstanceId !== undefined && object.lastInstanceId !== null) {
            message.lastInstanceId = object.lastInstanceId;
        }
        else {
            message.lastInstanceId = long_1.default.UZERO;
        }
        if (object.codes !== undefined && object.codes !== null) {
            for (const e of object.codes) {
                message.codes.push(exports.Code.fromPartial(e));
            }
        }
        if (object.contracts !== undefined && object.contracts !== null) {
            for (const e of object.contracts) {
                message.contracts.push(exports.Contract.fromPartial(e));
            }
        }
        return message;
    },
};
const baseModel = {};
exports.Model = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.key.length !== 0) {
            writer.uint32(10).bytes(message.key);
        }
        if (message.value.length !== 0) {
            writer.uint32(18).bytes(message.value);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = Object.assign({}, baseModel);
        message.key = new Uint8Array();
        message.value = new Uint8Array();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.key = reader.bytes();
                    break;
                case 2:
                    message.value = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const message = Object.assign({}, baseModel);
        message.key = new Uint8Array();
        message.value = new Uint8Array();
        if (object.key !== undefined && object.key !== null) {
            message.key = bytesFromBase64(object.key);
        }
        if (object.value !== undefined && object.value !== null) {
            message.value = bytesFromBase64(object.value);
        }
        return message;
    },
    toJSON(message) {
        const obj = {};
        message.key !== undefined &&
            (obj.key = base64FromBytes(message.key !== undefined ? message.key : new Uint8Array()));
        message.value !== undefined &&
            (obj.value = base64FromBytes(message.value !== undefined ? message.value : new Uint8Array()));
        return obj;
    },
    fromPartial(object) {
        const message = Object.assign({}, baseModel);
        if (object.key !== undefined && object.key !== null) {
            message.key = object.key;
        }
        else {
            message.key = new Uint8Array();
        }
        if (object.value !== undefined && object.value !== null) {
            message.value = object.value;
        }
        else {
            message.value = new Uint8Array();
        }
        return message;
    },
};
const baseCode = {};
exports.Code = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.codeInfo !== undefined) {
            wasm_1.CodeInfo.encode(message.codeInfo, writer.uint32(10).fork()).ldelim();
        }
        if (message.codeBytes.length !== 0) {
            writer.uint32(18).bytes(message.codeBytes);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = Object.assign({}, baseCode);
        message.codeBytes = new Uint8Array();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.codeInfo = wasm_1.CodeInfo.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.codeBytes = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const message = Object.assign({}, baseCode);
        message.codeBytes = new Uint8Array();
        if (object.codeInfo !== undefined && object.codeInfo !== null) {
            message.codeInfo = wasm_1.CodeInfo.fromJSON(object.codeInfo);
        }
        else {
            message.codeInfo = undefined;
        }
        if (object.codeBytes !== undefined && object.codeBytes !== null) {
            message.codeBytes = bytesFromBase64(object.codeBytes);
        }
        return message;
    },
    toJSON(message) {
        const obj = {};
        message.codeInfo !== undefined &&
            (obj.codeInfo = message.codeInfo ? wasm_1.CodeInfo.toJSON(message.codeInfo) : undefined);
        message.codeBytes !== undefined &&
            (obj.codeBytes = base64FromBytes(message.codeBytes !== undefined ? message.codeBytes : new Uint8Array()));
        return obj;
    },
    fromPartial(object) {
        const message = Object.assign({}, baseCode);
        if (object.codeInfo !== undefined && object.codeInfo !== null) {
            message.codeInfo = wasm_1.CodeInfo.fromPartial(object.codeInfo);
        }
        else {
            message.codeInfo = undefined;
        }
        if (object.codeBytes !== undefined && object.codeBytes !== null) {
            message.codeBytes = object.codeBytes;
        }
        else {
            message.codeBytes = new Uint8Array();
        }
        return message;
    },
};
const baseContract = {};
exports.Contract = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.contractInfo !== undefined) {
            wasm_1.ContractInfo.encode(message.contractInfo, writer.uint32(10).fork()).ldelim();
        }
        for (const v of message.contractStore) {
            exports.Model.encode(v, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = Object.assign({}, baseContract);
        message.contractStore = [];
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.contractInfo = wasm_1.ContractInfo.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.contractStore.push(exports.Model.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const message = Object.assign({}, baseContract);
        message.contractStore = [];
        if (object.contractInfo !== undefined && object.contractInfo !== null) {
            message.contractInfo = wasm_1.ContractInfo.fromJSON(object.contractInfo);
        }
        else {
            message.contractInfo = undefined;
        }
        if (object.contractStore !== undefined && object.contractStore !== null) {
            for (const e of object.contractStore) {
                message.contractStore.push(exports.Model.fromJSON(e));
            }
        }
        return message;
    },
    toJSON(message) {
        const obj = {};
        message.contractInfo !== undefined &&
            (obj.contractInfo = message.contractInfo ? wasm_1.ContractInfo.toJSON(message.contractInfo) : undefined);
        if (message.contractStore) {
            obj.contractStore = message.contractStore.map((e) => (e ? exports.Model.toJSON(e) : undefined));
        }
        else {
            obj.contractStore = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = Object.assign({}, baseContract);
        message.contractStore = [];
        if (object.contractInfo !== undefined && object.contractInfo !== null) {
            message.contractInfo = wasm_1.ContractInfo.fromPartial(object.contractInfo);
        }
        else {
            message.contractInfo = undefined;
        }
        if (object.contractStore !== undefined && object.contractStore !== null) {
            for (const e of object.contractStore) {
                message.contractStore.push(exports.Model.fromPartial(e));
            }
        }
        return message;
    },
};
var globalThis = (() => {
    if (typeof globalThis !== "undefined")
        return globalThis;
    if (typeof self !== "undefined")
        return self;
    if (typeof window !== "undefined")
        return window;
    if (typeof global !== "undefined")
        return global;
    throw "Unable to locate global object";
})();
const atob = globalThis.atob || ((b64) => globalThis.Buffer.from(b64, "base64").toString("binary"));
function bytesFromBase64(b64) {
    const bin = atob(b64);
    const arr = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; ++i) {
        arr[i] = bin.charCodeAt(i);
    }
    return arr;
}
const btoa = globalThis.btoa || ((bin) => globalThis.Buffer.from(bin, "binary").toString("base64"));
function base64FromBytes(arr) {
    const bin = [];
    for (const byte of arr) {
        bin.push(String.fromCharCode(byte));
    }
    return btoa(bin.join(""));
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
