// @hop-protocol/core/contracts/factories/ArbitrumL2ERC20Bridge__factory.js
"use strict";
// @ts-nocheck
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArbitrumL2ERC20Bridge__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "address",
                name: "_l2Messenger",
                type: "address",
            },
            {
                internalType: "address",
                name: "_l1ERC20BridgeAddress",
                type: "address",
            },
        ],
        stateMutability: "nonpayable",
        type: "constructor",
    },
    {
        inputs: [],
        name: "l1ERC20BridgeAddress",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "l2Messenger",
        outputs: [
            {
                internalType: "contract IArbSys",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "_l1TokenAddress",
                type: "address",
            },
            {
                internalType: "address",
                name: "_l2TokenAddress",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "_amount",
                type: "uint256",
            },
        ],
        name: "withdraw",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class ArbitrumL2ERC20Bridge__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.ArbitrumL2ERC20Bridge__factory = ArbitrumL2ERC20Bridge__factory;
ArbitrumL2ERC20Bridge__factory.abi = _abi;
