// @hop-protocol/core/contracts/factories/IGlobalInbox__factory.js
"use strict";
// @ts-nocheck
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IGlobalInbox__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                indexed: false,
                internalType: "bytes",
                name: "data",
                type: "bytes",
            },
        ],
        name: "BuddyContractDeployed",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address",
                name: "data",
                type: "address",
            },
        ],
        name: "BuddyContractPair",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "chain",
                type: "address",
            },
            {
                indexed: true,
                internalType: "uint8",
                name: "kind",
                type: "uint8",
            },
            {
                indexed: true,
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "inboxSeqNum",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "bytes",
                name: "data",
                type: "bytes",
            },
        ],
        name: "MessageDelivered",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "chain",
                type: "address",
            },
            {
                indexed: true,
                internalType: "uint8",
                name: "kind",
                type: "uint8",
            },
            {
                indexed: true,
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "inboxSeqNum",
                type: "uint256",
            },
        ],
        name: "MessageDeliveredFromOrigin",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "chain",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "maxGas",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "gasPriceBid",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "payment",
                type: "uint256",
            },
            {
                internalType: "bytes",
                name: "contractData",
                type: "bytes",
            },
        ],
        name: "deployL2ContractPair",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "account",
                type: "address",
            },
        ],
        name: "getInbox",
        outputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes",
                name: "messageData",
                type: "bytes",
            },
        ],
        name: "sendInitializationMessage",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "chain",
                type: "address",
            },
            {
                internalType: "bytes",
                name: "messageData",
                type: "bytes",
            },
        ],
        name: "sendL2Message",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes",
                name: "_messages",
                type: "bytes",
            },
            {
                internalType: "uint256",
                name: "initialMaxSendCount",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "finalMaxSendCount",
                type: "uint256",
            },
        ],
        name: "sendMessages",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class IGlobalInbox__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IGlobalInbox__factory = IGlobalInbox__factory;
IGlobalInbox__factory.abi = _abi;
