// @hop-protocol/core/contracts/factories/MockBridge__factory.js
"use strict";
// @ts-nocheck
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockBridge__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "address[]",
                name: "_bonders",
                type: "address[]",
            },
        ],
        stateMutability: "nonpayable",
        type: "constructor",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "newBonder",
                type: "address",
            },
        ],
        name: "BonderAdded",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "previousBonder",
                type: "address",
            },
        ],
        name: "BonderRemoved",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "bonder",
                type: "address",
            },
            {
                indexed: true,
                internalType: "bytes32",
                name: "rootHash",
                type: "bytes32",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "totalBondsSettled",
                type: "uint256",
            },
        ],
        name: "MultipleWithdrawalsSettled",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "account",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
        ],
        name: "Stake",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "bytes32",
                name: "rootHash",
                type: "bytes32",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "totalAmount",
                type: "uint256",
            },
        ],
        name: "TransferRootSet",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "account",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
        ],
        name: "Unstake",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "bonder",
                type: "address",
            },
            {
                indexed: true,
                internalType: "bytes32",
                name: "transferId",
                type: "bytes32",
            },
            {
                indexed: true,
                internalType: "bytes32",
                name: "rootHash",
                type: "bytes32",
            },
        ],
        name: "WithdrawalBondSettled",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "bytes32",
                name: "transferId",
                type: "bytes32",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
        ],
        name: "WithdrawalBonded",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "bytes32",
                name: "transferId",
                type: "bytes32",
            },
            {
                indexed: true,
                internalType: "address",
                name: "recipient",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "bytes32",
                name: "transferNonce",
                type: "bytes32",
            },
        ],
        name: "Withdrew",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "bonder",
                type: "address",
            },
        ],
        name: "addBonder",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "recipient",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
            {
                internalType: "bytes32",
                name: "transferNonce",
                type: "bytes32",
            },
            {
                internalType: "uint256",
                name: "bonderFee",
                type: "uint256",
            },
        ],
        name: "bondWithdrawal",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "bonder",
                type: "address",
            },
            {
                internalType: "bytes32",
                name: "transferId",
                type: "bytes32",
            },
        ],
        name: "getBondedWithdrawalAmount",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "getChainId",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "bonder",
                type: "address",
            },
        ],
        name: "getCredit",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "bonder",
                type: "address",
            },
        ],
        name: "getDebitAndAdditionalDebit",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "maybeBonder",
                type: "address",
            },
        ],
        name: "getIsBonder",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "bonder",
                type: "address",
            },
        ],
        name: "getRawDebit",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "chainId",
                type: "uint256",
            },
            {
                internalType: "address",
                name: "recipient",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
            {
                internalType: "bytes32",
                name: "transferNonce",
                type: "bytes32",
            },
            {
                internalType: "uint256",
                name: "bonderFee",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "amountOutMin",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "deadline",
                type: "uint256",
            },
        ],
        name: "getTransferId",
        outputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        stateMutability: "pure",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "rootHash",
                type: "bytes32",
            },
            {
                internalType: "uint256",
                name: "totalAmount",
                type: "uint256",
            },
        ],
        name: "getTransferRoot",
        outputs: [
            {
                components: [
                    {
                        internalType: "uint256",
                        name: "total",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "amountWithdrawn",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "createdAt",
                        type: "uint256",
                    },
                ],
                internalType: "struct Bridge.TransferRoot",
                name: "",
                type: "tuple",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "rootHash",
                type: "bytes32",
            },
            {
                internalType: "uint256",
                name: "totalAmount",
                type: "uint256",
            },
        ],
        name: "getTransferRootId",
        outputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        stateMutability: "pure",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "transferId",
                type: "bytes32",
            },
        ],
        name: "isTransferIdSpent",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "bonder",
                type: "address",
            },
        ],
        name: "removeBonder",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "rootHash",
                type: "bytes32",
            },
            {
                internalType: "uint256",
                name: "originalAmount",
                type: "uint256",
            },
            {
                internalType: "address",
                name: "recipient",
                type: "address",
            },
        ],
        name: "rescueTransferRoot",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "bonder",
                type: "address",
            },
            {
                internalType: "bytes32",
                name: "transferId",
                type: "bytes32",
            },
            {
                internalType: "bytes32",
                name: "rootHash",
                type: "bytes32",
            },
            {
                internalType: "uint256",
                name: "transferRootTotalAmount",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "transferIdTreeIndex",
                type: "uint256",
            },
            {
                internalType: "bytes32[]",
                name: "siblings",
                type: "bytes32[]",
            },
            {
                internalType: "uint256",
                name: "totalLeaves",
                type: "uint256",
            },
        ],
        name: "settleBondedWithdrawal",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "bonder",
                type: "address",
            },
            {
                internalType: "bytes32[]",
                name: "transferIds",
                type: "bytes32[]",
            },
            {
                internalType: "uint256",
                name: "totalAmount",
                type: "uint256",
            },
        ],
        name: "settleBondedWithdrawals",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "bonder",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
        ],
        name: "stake",
        outputs: [],
        stateMutability: "payable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
        ],
        name: "unstake",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "recipient",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
            {
                internalType: "bytes32",
                name: "transferNonce",
                type: "bytes32",
            },
            {
                internalType: "uint256",
                name: "bonderFee",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "amountOutMin",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "deadline",
                type: "uint256",
            },
            {
                internalType: "bytes32",
                name: "rootHash",
                type: "bytes32",
            },
            {
                internalType: "uint256",
                name: "transferRootTotalAmount",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "transferIdTreeIndex",
                type: "uint256",
            },
            {
                internalType: "bytes32[]",
                name: "siblings",
                type: "bytes32[]",
            },
            {
                internalType: "uint256",
                name: "totalLeaves",
                type: "uint256",
            },
        ],
        name: "withdraw",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class MockBridge__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.MockBridge__factory = MockBridge__factory;
MockBridge__factory.abi = _abi;
