// @hop-protocol/core/contracts/factories/MockEthERC20Bridge__factory.js
"use strict";
// @ts-nocheck
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockEthERC20Bridge__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "address",
                name: "erc20",
                type: "address",
            },
            {
                internalType: "address",
                name: "destination",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "maxSubmissionCost",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "maxGas",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "gasPriceBid",
                type: "uint256",
            },
            {
                internalType: "bytes",
                name: "callHookData",
                type: "bytes",
            },
        ],
        name: "depositAsERC20",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class MockEthERC20Bridge__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.MockEthERC20Bridge__factory = MockEthERC20Bridge__factory;
MockEthERC20Bridge__factory.abi = _abi;
