// @hop-protocol/core/metadata/assets/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WBTCImage = exports.USDTImage = exports.USDCImage = exports.sETHImage = exports.sBTCImage = exports.DAIImage = exports.ARBImage = exports.XDAIImage = exports.MATICImage = exports.ETHImage = void 0;
var ETH_1 = require("./ETH");
Object.defineProperty(exports, "ETHImage", { enumerable: true, get: function () { return __importDefault(ETH_1).default; } });
var MATIC_1 = require("./MATIC");
Object.defineProperty(exports, "MATICImage", { enumerable: true, get: function () { return __importDefault(MATIC_1).default; } });
var XDAI_1 = require("./XDAI");
Object.defineProperty(exports, "XDAIImage", { enumerable: true, get: function () { return __importDefault(XDAI_1).default; } });
var ARB_1 = require("./ARB");
Object.defineProperty(exports, "ARBImage", { enumerable: true, get: function () { return __importDefault(ARB_1).default; } });
var DAI_1 = require("./DAI");
Object.defineProperty(exports, "DAIImage", { enumerable: true, get: function () { return __importDefault(DAI_1).default; } });
var sBTC_1 = require("./sBTC");
Object.defineProperty(exports, "sBTCImage", { enumerable: true, get: function () { return __importDefault(sBTC_1).default; } });
var sETH_1 = require("./sETH");
Object.defineProperty(exports, "sETHImage", { enumerable: true, get: function () { return __importDefault(sETH_1).default; } });
var USDC_1 = require("./USDC");
Object.defineProperty(exports, "USDCImage", { enumerable: true, get: function () { return __importDefault(USDC_1).default; } });
var USDT_1 = require("./USDT");
Object.defineProperty(exports, "USDTImage", { enumerable: true, get: function () { return __importDefault(USDT_1).default; } });
var WBTC_1 = require("./WBTC");
Object.defineProperty(exports, "WBTCImage", { enumerable: true, get: function () { return __importDefault(WBTC_1).default; } });
