// @hop-protocol/core/networks/kovan.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.networks = void 0;
const metadata_1 = require("../metadata");
exports.networks = {
    ethereum: {
        name: metadata_1.chains.ethereum.name,
        networkId: 42,
        publicRpcUrl: 'https://kovan.infura.io/v3/84842078b09946638c03157f83405213',
        explorerUrls: ['https://kovan.etherscan.io'],
        waitConfirmations: 1
    },
    optimism: {
        name: metadata_1.chains.optimism.name,
        networkId: 69,
        publicRpcUrl: 'https://kovan.optimism.io',
        explorerUrls: ['https://kovan-optimistic.etherscan.io'],
        nativeBridgeUrl: 'https://gateway.optimism.io/welcome',
        waitConfirmations: 1
    },
    gnosis: {
        name: metadata_1.chains.gnosis.name,
        networkId: 77,
        publicRpcUrl: 'https://sokol.poa.network',
        explorerUrls: ['https://blockscout.com/poa/sokol'],
        nativeBridgeUrl: 'https://omni.xdaichain.com/',
        waitConfirmations: 1
    }
};
