// @hop-protocol/core/networks/mainnet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.networks = void 0;
const metadata_1 = require("../metadata");
exports.networks = {
    ethereum: {
        name: metadata_1.chains.ethereum.name,
        networkId: 1,
        publicRpcUrl: 'https://mainnet.infura.io/v3/84842078b09946638c03157f83405213',
        explorerUrls: ['https://etherscan.io'],
        waitConfirmations: 12
    },
    arbitrum: {
        name: metadata_1.chains.arbitrum.name,
        networkId: 42161,
        publicRpcUrl: 'https://arb1.arbitrum.io/rpc',
        explorerUrls: ['https://arbiscan.io/'],
        nativeBridgeUrl: 'https://bridge.arbitrum.io/',
        waitConfirmations: 20
    },
    optimism: {
        name: metadata_1.chains.optimism.name,
        networkId: 10,
        publicRpcUrl: 'https://mainnet.optimism.io',
        explorerUrls: ['https://optimistic.etherscan.io'],
        nativeBridgeUrl: 'https://gateway.optimism.io/welcome',
        waitConfirmations: 1
    },
    gnosis: {
        name: metadata_1.chains.gnosis.name,
        networkId: 100,
        publicRpcUrl: 'https://rpc.gnosischain.com/',
        explorerUrls: ['https://blockscout.com/xdai/mainnet'],
        nativeBridgeUrl: 'https://omni.xdaichain.com/',
        waitConfirmations: 12
    },
    polygon: {
        name: metadata_1.chains.polygon.name,
        networkId: 137,
        publicRpcUrl: 'https://polygon-rpc.com',
        explorerUrls: ['https://polygonscan.com'],
        nativeBridgeUrl: 'https://wallet.matic.network/bridge/',
        waitConfirmations: 256
    }
};
