// @ledgerhq/hw-app-btc/lib-es/BtcNew.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { crypto } from "bitcoinjs-lib";
import { pointCompress } from "tiny-secp256k1";
import { getXpubComponents, hardenedPathOf, pathArrayToString, pathStringToArray, pubkeyFromXpub, } from "./bip32";
import { BufferReader } from "./buffertools";
import { p2pkh, p2tr, p2wpkh, p2wpkhWrapped, } from "./newops/accounttype";
import { createKey, WalletPolicy, } from "./newops/policy";
import { extract } from "./newops/psbtExtractor";
import { finalize } from "./newops/psbtFinalizer";
import { psbtIn, PsbtV2 } from "./newops/psbtv2";
import { serializeTransaction } from "./serializeTransaction";
/**
 * This class implements the same interface as BtcOld (formerly
 * named Btc), but interacts with Bitcoin hardware app version 2+
 * which uses a totally new APDU protocol. This new
 * protocol is documented at
 * https://github.com/LedgerHQ/app-bitcoin-new/blob/master/doc/bitcoin.md
 *
 * Since the interface must remain compatible with BtcOld, the methods
 * of this class are quite clunky, because it needs to adapt legacy
 * input data into the PSBT process. In the future, a new interface should
 * be developed that exposes PSBT to the outer world, which would render
 * a much cleaner implementation.
 */
var BtcNew = /** @class */ (function () {
    function BtcNew(client) {
        this.client = client;
    }
    /**
     * This is a new method that allow users to get an xpub at a standard path.
     * Standard paths are described at
     * https://github.com/LedgerHQ/app-bitcoin-new/blob/master/doc/bitcoin.md#description
     *
     * This boils down to paths (N=0 for Bitcoin, N=1 for Testnet):
     * M/44'/N'/x'/**
     * M/48'/N'/x'/y'/**
     * M/49'/N'/x'/**
     * M/84'/N'/x'/**
     * M/86'/N'/x'/**
     *
     * The method was added because of added security in the hardware app v2+. The
     * new hardware app will allow export of any xpub up to and including the
     * deepest hardened key of standard derivation paths, whereas the old app
     * would allow export of any key.
     *
     * This caused an issue for callers of this class, who only had
     * getWalletPublicKey() to call which means they have to constuct xpub
     * themselves:
     *
     * Suppose a user of this class wants to create an account xpub on a standard
     * path, M/44'/0'/Z'. The user must get the parent key fingerprint (see BIP32)
     * by requesting the parent key M/44'/0'. The new app won't allow that, because
     * it only allows exporting deepest level hardened path. So the options are to
     * allow requesting M/44'/0' from the app, or to add a new function
     * "getWalletXpub".
     *
     * We opted for adding a new function, which can greatly simplify client code.
     */
    BtcNew.prototype.getWalletXpub = function (_a) {
        var path = _a.path, xpubVersion = _a.xpubVersion;
        return __awaiter(this, void 0, void 0, function () {
            var pathElements, xpub, xpubComponents;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        pathElements = pathStringToArray(path);
                        return [4 /*yield*/, this.client.getExtendedPubkey(false, pathElements)];
                    case 1:
                        xpub = _b.sent();
                        xpubComponents = getXpubComponents(xpub);
                        if (xpubComponents.version != xpubVersion) {
                            throw new Error("Expected xpub version ".concat(xpubVersion, " doesn't match the xpub version from the device ").concat(xpubComponents.version));
                        }
                        return [2 /*return*/, xpub];
                }
            });
        });
    };
    /**
     * This method returns a public key, a bitcoin address, and and a chaincode
     * for a specific derivation path.
     *
     * Limitation: If the path is not a leaf node of a standard path, the address
     * will be the empty string "", see this.getWalletAddress() for details.
     */
    BtcNew.prototype.getWalletPublicKey = function (path, opts) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function () {
            var pathElements, xpub, display, address, components, uncompressedPubkey;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        if (!isPathNormal(path)) {
                            throw Error("non-standard path: ".concat(path));
                        }
                        pathElements = pathStringToArray(path);
                        return [4 /*yield*/, this.client.getExtendedPubkey(false, pathElements)];
                    case 1:
                        xpub = _c.sent();
                        display = (_a = opts === null || opts === void 0 ? void 0 : opts.verify) !== null && _a !== void 0 ? _a : false;
                        return [4 /*yield*/, this.getWalletAddress(pathElements, descrTemplFrom((_b = opts === null || opts === void 0 ? void 0 : opts.format) !== null && _b !== void 0 ? _b : "legacy"), display)];
                    case 2:
                        address = _c.sent();
                        components = getXpubComponents(xpub);
                        uncompressedPubkey = Buffer.from(pointCompress(components.pubkey, false));
                        return [2 /*return*/, {
                                publicKey: uncompressedPubkey.toString("hex"),
                                bitcoinAddress: address,
                                chainCode: components.chaincode.toString("hex")
                            }];
                }
            });
        });
    };
    /**
     * Get an address for the specified path.
     *
     * If display is true, we must get the address from the device, which would require
     * us to determine WalletPolicy. This requires two *extra* queries to the device, one
     * for the account xpub and one for master key fingerprint.
     *
     * If display is false we *could* generate the address ourselves, but chose to
     * get it from the device to save development time. However, it shouldn't take
     * too much time to implement local address generation.
     *
     * Moreover, if the path is not for a leaf, ie accountPath+/X/Y, there is no
     * way to get the address from the device. In this case we have to create it
     * ourselves, but we don't at this time, and instead return an empty ("") address.
     */
    BtcNew.prototype.getWalletAddress = function (pathElements, descrTempl, display) {
        return __awaiter(this, void 0, void 0, function () {
            var accountPath, accountXpub, masterFingerprint, policy, changeAndIndex;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        accountPath = hardenedPathOf(pathElements);
                        if (accountPath.length + 2 != pathElements.length) {
                            return [2 /*return*/, ""];
                        }
                        return [4 /*yield*/, this.client.getExtendedPubkey(false, accountPath)];
                    case 1:
                        accountXpub = _a.sent();
                        return [4 /*yield*/, this.client.getMasterFingerprint()];
                    case 2:
                        masterFingerprint = _a.sent();
                        policy = new WalletPolicy(descrTempl, createKey(masterFingerprint, accountPath, accountXpub));
                        changeAndIndex = pathElements.slice(-2, pathElements.length);
                        return [2 /*return*/, this.client.getWalletAddress(policy, Buffer.alloc(32, 0), changeAndIndex[0], changeAndIndex[1], display)];
                }
            });
        });
    };
    /**
     * Build and sign a transaction. See Btc.createPaymentTransaction for
     * details on how to use this method.
     *
     * This method will convert the legacy arguments, CreateTransactionArg, into
     * a psbt which is finally signed and finalized, and the extracted fully signed
     * transaction is returned.
     */
    BtcNew.prototype.createPaymentTransaction = function (arg) {
        return __awaiter(this, void 0, void 0, function () {
            var inputCount, psbt, masterFp, accountType, notifyCount, progress, accountXpub, accountPath, i, pathElems, outputsConcat, outputsBufferReader, outputCount, changeData, changeFound, i, amount, outputScript, isChange, changePath, pubkey, key, p, firstSigned, progressCallback, serializedTx;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        inputCount = arg.inputs.length;
                        if (inputCount == 0) {
                            throw Error("No inputs");
                        }
                        psbt = new PsbtV2();
                        return [4 /*yield*/, this.client.getMasterFingerprint()];
                    case 1:
                        masterFp = _a.sent();
                        accountType = accountTypeFromArg(arg, psbt, masterFp);
                        if (arg.lockTime != undefined) {
                            // The signer will assume locktime 0 if unset
                            psbt.setGlobalFallbackLocktime(arg.lockTime);
                        }
                        psbt.setGlobalInputCount(inputCount);
                        psbt.setGlobalPsbtVersion(2);
                        psbt.setGlobalTxVersion(2);
                        notifyCount = 0;
                        progress = function () {
                            if (!arg.onDeviceStreaming)
                                return;
                            arg.onDeviceStreaming({
                                total: 2 * inputCount,
                                index: notifyCount,
                                progress: ++notifyCount / (2 * inputCount)
                            });
                        };
                        accountXpub = "";
                        accountPath = [];
                        i = 0;
                        _a.label = 2;
                    case 2:
                        if (!(i < inputCount)) return [3 /*break*/, 7];
                        progress();
                        pathElems = pathStringToArray(arg.associatedKeysets[i]);
                        if (!(accountXpub == "")) return [3 /*break*/, 4];
                        // We assume all inputs belong to the same account so we set
                        // the account xpub and path based on the first input.
                        accountPath = pathElems.slice(0, -2);
                        return [4 /*yield*/, this.client.getExtendedPubkey(false, accountPath)];
                    case 3:
                        accountXpub = _a.sent();
                        _a.label = 4;
                    case 4: return [4 /*yield*/, this.setInput(psbt, i, arg.inputs[i], pathElems, accountType, masterFp, arg.sigHashType)];
                    case 5:
                        _a.sent();
                        _a.label = 6;
                    case 6:
                        i++;
                        return [3 /*break*/, 2];
                    case 7:
                        outputsConcat = Buffer.from(arg.outputScriptHex, "hex");
                        outputsBufferReader = new BufferReader(outputsConcat);
                        outputCount = outputsBufferReader.readVarInt();
                        psbt.setGlobalOutputCount(outputCount);
                        return [4 /*yield*/, this.outputScriptAt(accountPath, accountType, arg.changePath)];
                    case 8:
                        changeData = _a.sent();
                        changeFound = !changeData;
                        for (i = 0; i < outputCount; i++) {
                            amount = Number(outputsBufferReader.readUInt64());
                            outputScript = outputsBufferReader.readVarSlice();
                            psbt.setOutputAmount(i, amount);
                            psbt.setOutputScript(i, outputScript);
                            isChange = changeData && outputScript.equals(changeData === null || changeData === void 0 ? void 0 : changeData.cond.scriptPubKey);
                            if (isChange) {
                                changeFound = true;
                                changePath = pathStringToArray(arg.changePath);
                                pubkey = changeData.pubkey;
                                accountType.setOwnOutput(i, changeData.cond, [pubkey], [changePath]);
                            }
                        }
                        if (!changeFound) {
                            throw new Error("Change script not found among outputs! " +
                                (changeData === null || changeData === void 0 ? void 0 : changeData.cond.scriptPubKey.toString("hex")));
                        }
                        key = createKey(masterFp, accountPath, accountXpub);
                        p = new WalletPolicy(accountType.getDescriptorTemplate(), key);
                        // This is cheating, because it's not actually requested on the
                        // device yet, but it will be, soonish.
                        if (arg.onDeviceSignatureRequested)
                            arg.onDeviceSignatureRequested();
                        firstSigned = false;
                        progressCallback = function () {
                            if (!firstSigned) {
                                firstSigned = true;
                                arg.onDeviceSignatureGranted && arg.onDeviceSignatureGranted();
                            }
                            progress();
                        };
                        return [4 /*yield*/, this.signPsbt(psbt, p, progressCallback)];
                    case 9:
                        _a.sent();
                        finalize(psbt);
                        serializedTx = extract(psbt);
                        return [2 /*return*/, serializedTx.toString("hex")];
                }
            });
        });
    };
    /**
     * Calculates an output script along with public key and possible redeemScript
     * from a path and accountType. The accountPath must be a prefix of path.
     *
     * @returns an object with output script (property "script"), redeemScript (if
     * wrapped p2wpkh), and pubkey at provided path. The values of these three
     * properties depend on the accountType used.
     */
    BtcNew.prototype.outputScriptAt = function (accountPath, accountType, path) {
        return __awaiter(this, void 0, void 0, function () {
            var pathElems, i, xpub, pubkey, cond;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!path)
                            return [2 /*return*/, undefined];
                        pathElems = pathStringToArray(path);
                        // Make sure path is in our account, otherwise something fishy is probably
                        // going on.
                        for (i = 0; i < accountPath.length; i++) {
                            if (accountPath[i] != pathElems[i]) {
                                throw new Error("Path ".concat(path, " not in account ").concat(pathArrayToString(accountPath)));
                            }
                        }
                        return [4 /*yield*/, this.client.getExtendedPubkey(false, pathElems)];
                    case 1:
                        xpub = _a.sent();
                        pubkey = pubkeyFromXpub(xpub);
                        cond = accountType.spendingCondition([pubkey]);
                        return [2 /*return*/, { cond: cond, pubkey: pubkey }];
                }
            });
        });
    };
    /**
     * Adds relevant data about an input to the psbt. This includes sequence,
     * previous txid, output index, spent UTXO, redeem script for wrapped p2wpkh,
     * public key and its derivation path.
     */
    BtcNew.prototype.setInput = function (psbt, i, input, pathElements, accountType, masterFP, sigHashType) {
        return __awaiter(this, void 0, void 0, function () {
            var inputTx, spentOutputIndex, redeemScript, sequence, inputTxBuffer, inputTxid, xpubBase58, pubkey, spentTxOutput, spendCondition, spentOutput;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        inputTx = input[0];
                        spentOutputIndex = input[1];
                        redeemScript = input[2] ? Buffer.from(input[2], "hex") : undefined;
                        sequence = input[3];
                        if (sequence != undefined) {
                            psbt.setInputSequence(i, sequence);
                        }
                        if (sigHashType != undefined) {
                            psbt.setInputSighashType(i, sigHashType);
                        }
                        inputTxBuffer = serializeTransaction(inputTx, true);
                        inputTxid = crypto.hash256(inputTxBuffer);
                        return [4 /*yield*/, this.client.getExtendedPubkey(false, pathElements)];
                    case 1:
                        xpubBase58 = _a.sent();
                        pubkey = pubkeyFromXpub(xpubBase58);
                        if (!inputTx.outputs)
                            throw Error("Missing outputs array in transaction to sign");
                        spentTxOutput = inputTx.outputs[spentOutputIndex];
                        spendCondition = {
                            scriptPubKey: spentTxOutput.script,
                            redeemScript: redeemScript
                        };
                        spentOutput = { cond: spendCondition, amount: spentTxOutput.amount };
                        accountType.setInput(i, inputTxBuffer, spentOutput, [pubkey], [pathElements]);
                        psbt.setInputPreviousTxId(i, inputTxid);
                        psbt.setInputOutputIndex(i, spentOutputIndex);
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * This implements the "Signer" role of the BIP370 transaction signing
     * process.
     *
     * It ssks the hardware device to sign the a psbt using the specified wallet
     * policy. This method assumes BIP32 derived keys are used for all inputs, see
     * comment in-line. The signatures returned from the hardware device is added
     * to the appropriate input fields of the PSBT.
     */
    BtcNew.prototype.signPsbt = function (psbt, walletPolicy, progressCallback) {
        return __awaiter(this, void 0, void 0, function () {
            var sigs;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.client.signPsbt(psbt, walletPolicy, Buffer.alloc(32, 0), progressCallback)];
                    case 1:
                        sigs = _a.sent();
                        sigs.forEach(function (v, k) {
                            // Note: Looking at BIP32 derivation does not work in the generic case,
                            // since some inputs might not have a BIP32-derived pubkey.
                            var pubkeys = psbt.getInputKeyDatas(k, psbtIn.BIP32_DERIVATION);
                            var pubkey;
                            if (pubkeys.length != 1) {
                                // No legacy BIP32_DERIVATION, assume we're using taproot.
                                pubkey = psbt.getInputKeyDatas(k, psbtIn.TAP_BIP32_DERIVATION);
                                if (pubkey.length == 0) {
                                    throw Error("Missing pubkey derivation for input ".concat(k));
                                }
                                psbt.setInputTapKeySig(k, v);
                            }
                            else {
                                pubkey = pubkeys[0];
                                psbt.setInputPartialSig(k, pubkey, v);
                            }
                        });
                        return [2 /*return*/];
                }
            });
        });
    };
    return BtcNew;
}());
export default BtcNew;
function descrTemplFrom(addressFormat) {
    if (addressFormat == "legacy")
        return "pkh(@0)";
    if (addressFormat == "p2sh")
        return "sh(wpkh(@0))";
    if (addressFormat == "bech32")
        return "wpkh(@0)";
    if (addressFormat == "bech32m")
        return "tr(@0)";
    throw new Error("Unsupported address format " + addressFormat);
}
function accountTypeFromArg(arg, psbt, masterFp) {
    if (arg.additionals.includes("bech32m"))
        return new p2tr(psbt, masterFp);
    if (arg.additionals.includes("bech32"))
        return new p2wpkh(psbt, masterFp);
    if (arg.segwit)
        return new p2wpkhWrapped(psbt, masterFp);
    return new p2pkh(psbt, masterFp);
}
/*
  The new protocol only allows standard path.
  Standard paths are (currently):
  M/44'/(1|0)'/X'
  M/49'/(1|0)'/X'
  M/84'/(1|0)'/X'
  M/86'/(1|0)'/X'
  M/48'/(1|0)'/X'/Y'
  followed by "", "(0|1)", or "(0|1)/b", where a and b are
  non-hardened. For example, the following paths are standard
  M/48'/1'/99'/7'
  M/86'/1'/99'/0
  M/48'/0'/99'/7'/1/17
  The following paths are non-standard
  M/48'/0'/99'           // Not deepest hardened path
  M/48'/0'/99'/7'/1/17/2 // Too many non-hardened derivation steps
  M/199'/0'/1'/0/88      // Not a known purpose 199
  M/86'/1'/99'/2         // Change path item must be 0 or 1
*/
function isPathNormal(path) {
    //path is not deepest hardened node of a standard path or deeper, use BtcOld
    var h = 0x80000000; //HARDENED from bip32
    var pathElems = pathStringToArray(path);
    var hard = function (n) { return n >= h; };
    var soft = function (n) { return n === undefined || n < h; };
    var change = function (n) {
        return n === undefined || n === 0 || n === 1;
    };
    if (pathElems.length >= 3 &&
        pathElems.length <= 5 &&
        [44 + h, 49 + h, 84 + h, 86 + h].some(function (v) { return v == pathElems[0]; }) &&
        [0 + h, 1 + h].some(function (v) { return v == pathElems[1]; }) &&
        hard(pathElems[2]) &&
        change(pathElems[3]) &&
        soft(pathElems[4])) {
        return true;
    }
    if (pathElems.length >= 4 &&
        pathElems.length <= 6 &&
        48 + h == pathElems[0] &&
        [0 + h, 1 + h].some(function (v) { return v == pathElems[1]; }) &&
        hard(pathElems[2]) &&
        hard(pathElems[3]) &&
        change(pathElems[4]) &&
        soft(pathElems[5])) {
        return true;
    }
    return false;
}
