// @terra-money/legacy.proto/terra/market/v1beta1/genesis.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenesisState = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const market_1 = require("../../../terra/market/v1beta1/market");
exports.protobufPackage = "terra.market.v1beta1";
const baseGenesisState = {};
exports.GenesisState = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.params !== undefined) {
            market_1.Params.encode(message.params, writer.uint32(10).fork()).ldelim();
        }
        if (message.terraPoolDelta.length !== 0) {
            writer.uint32(18).bytes(message.terraPoolDelta);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = Object.assign({}, baseGenesisState);
        message.terraPoolDelta = new Uint8Array();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.params = market_1.Params.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.terraPoolDelta = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const message = Object.assign({}, baseGenesisState);
        message.terraPoolDelta = new Uint8Array();
        if (object.params !== undefined && object.params !== null) {
            message.params = market_1.Params.fromJSON(object.params);
        }
        else {
            message.params = undefined;
        }
        if (object.terraPoolDelta !== undefined && object.terraPoolDelta !== null) {
            message.terraPoolDelta = bytesFromBase64(object.terraPoolDelta);
        }
        return message;
    },
    toJSON(message) {
        const obj = {};
        message.params !== undefined && (obj.params = message.params ? market_1.Params.toJSON(message.params) : undefined);
        message.terraPoolDelta !== undefined &&
            (obj.terraPoolDelta = base64FromBytes(message.terraPoolDelta !== undefined ? message.terraPoolDelta : new Uint8Array()));
        return obj;
    },
    fromPartial(object) {
        const message = Object.assign({}, baseGenesisState);
        if (object.params !== undefined && object.params !== null) {
            message.params = market_1.Params.fromPartial(object.params);
        }
        else {
            message.params = undefined;
        }
        if (object.terraPoolDelta !== undefined && object.terraPoolDelta !== null) {
            message.terraPoolDelta = object.terraPoolDelta;
        }
        else {
            message.terraPoolDelta = new Uint8Array();
        }
        return message;
    },
};
var globalThis = (() => {
    if (typeof globalThis !== "undefined")
        return globalThis;
    if (typeof self !== "undefined")
        return self;
    if (typeof window !== "undefined")
        return window;
    if (typeof global !== "undefined")
        return global;
    throw "Unable to locate global object";
})();
const atob = globalThis.atob || ((b64) => globalThis.Buffer.from(b64, "base64").toString("binary"));
function bytesFromBase64(b64) {
    const bin = atob(b64);
    const arr = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; ++i) {
        arr[i] = bin.charCodeAt(i);
    }
    return arr;
}
const btoa = globalThis.btoa || ((bin) => globalThis.Buffer.from(bin, "binary").toString("base64"));
function base64FromBytes(arr) {
    const bin = [];
    for (const byte of arr) {
        bin.push(String.fromCharCode(byte));
    }
    return btoa(bin.join(""));
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
