// @terra-money/legacy.proto/terra/market/v1beta1/market.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "terra.market.v1beta1";
const baseParams = { poolRecoveryPeriod: long_1.default.UZERO };
exports.Params = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.basePool.length !== 0) {
            writer.uint32(10).bytes(message.basePool);
        }
        if (!message.poolRecoveryPeriod.isZero()) {
            writer.uint32(16).uint64(message.poolRecoveryPeriod);
        }
        if (message.minStabilitySpread.length !== 0) {
            writer.uint32(26).bytes(message.minStabilitySpread);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = Object.assign({}, baseParams);
        message.basePool = new Uint8Array();
        message.minStabilitySpread = new Uint8Array();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.basePool = reader.bytes();
                    break;
                case 2:
                    message.poolRecoveryPeriod = reader.uint64();
                    break;
                case 3:
                    message.minStabilitySpread = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const message = Object.assign({}, baseParams);
        message.basePool = new Uint8Array();
        message.minStabilitySpread = new Uint8Array();
        if (object.basePool !== undefined && object.basePool !== null) {
            message.basePool = bytesFromBase64(object.basePool);
        }
        if (object.poolRecoveryPeriod !== undefined && object.poolRecoveryPeriod !== null) {
            message.poolRecoveryPeriod = long_1.default.fromString(object.poolRecoveryPeriod);
        }
        else {
            message.poolRecoveryPeriod = long_1.default.UZERO;
        }
        if (object.minStabilitySpread !== undefined && object.minStabilitySpread !== null) {
            message.minStabilitySpread = bytesFromBase64(object.minStabilitySpread);
        }
        return message;
    },
    toJSON(message) {
        const obj = {};
        message.basePool !== undefined &&
            (obj.basePool = base64FromBytes(message.basePool !== undefined ? message.basePool : new Uint8Array()));
        message.poolRecoveryPeriod !== undefined &&
            (obj.poolRecoveryPeriod = (message.poolRecoveryPeriod || long_1.default.UZERO).toString());
        message.minStabilitySpread !== undefined &&
            (obj.minStabilitySpread = base64FromBytes(message.minStabilitySpread !== undefined ? message.minStabilitySpread : new Uint8Array()));
        return obj;
    },
    fromPartial(object) {
        const message = Object.assign({}, baseParams);
        if (object.basePool !== undefined && object.basePool !== null) {
            message.basePool = object.basePool;
        }
        else {
            message.basePool = new Uint8Array();
        }
        if (object.poolRecoveryPeriod !== undefined && object.poolRecoveryPeriod !== null) {
            message.poolRecoveryPeriod = object.poolRecoveryPeriod;
        }
        else {
            message.poolRecoveryPeriod = long_1.default.UZERO;
        }
        if (object.minStabilitySpread !== undefined && object.minStabilitySpread !== null) {
            message.minStabilitySpread = object.minStabilitySpread;
        }
        else {
            message.minStabilitySpread = new Uint8Array();
        }
        return message;
    },
};
var globalThis = (() => {
    if (typeof globalThis !== "undefined")
        return globalThis;
    if (typeof self !== "undefined")
        return self;
    if (typeof window !== "undefined")
        return window;
    if (typeof global !== "undefined")
        return global;
    throw "Unable to locate global object";
})();
const atob = globalThis.atob || ((b64) => globalThis.Buffer.from(b64, "base64").toString("binary"));
function bytesFromBase64(b64) {
    const bin = atob(b64);
    const arr = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; ++i) {
        arr[i] = bin.charCodeAt(i);
    }
    return arr;
}
const btoa = globalThis.btoa || ((bin) => globalThis.Buffer.from(bin, "binary").toString("base64"));
function base64FromBytes(arr) {
    const bin = [];
    for (const byte of arr) {
        bin.push(String.fromCharCode(byte));
    }
    return btoa(bin.join(""));
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
