// @terra-money/legacy.proto/terra/market/v1beta1/query.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryClientImpl = exports.QueryParamsResponse = exports.QueryParamsRequest = exports.QueryTerraPoolDeltaResponse = exports.QueryTerraPoolDeltaRequest = exports.QuerySwapResponse = exports.QuerySwapRequest = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const coin_1 = require("../../../cosmos/base/v1beta1/coin");
const market_1 = require("../../../terra/market/v1beta1/market");
exports.protobufPackage = "terra.market.v1beta1";
const baseQuerySwapRequest = { offerCoin: "", askDenom: "" };
exports.QuerySwapRequest = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.offerCoin !== "") {
            writer.uint32(10).string(message.offerCoin);
        }
        if (message.askDenom !== "") {
            writer.uint32(18).string(message.askDenom);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = Object.assign({}, baseQuerySwapRequest);
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.offerCoin = reader.string();
                    break;
                case 2:
                    message.askDenom = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const message = Object.assign({}, baseQuerySwapRequest);
        if (object.offerCoin !== undefined && object.offerCoin !== null) {
            message.offerCoin = String(object.offerCoin);
        }
        else {
            message.offerCoin = "";
        }
        if (object.askDenom !== undefined && object.askDenom !== null) {
            message.askDenom = String(object.askDenom);
        }
        else {
            message.askDenom = "";
        }
        return message;
    },
    toJSON(message) {
        const obj = {};
        message.offerCoin !== undefined && (obj.offerCoin = message.offerCoin);
        message.askDenom !== undefined && (obj.askDenom = message.askDenom);
        return obj;
    },
    fromPartial(object) {
        const message = Object.assign({}, baseQuerySwapRequest);
        if (object.offerCoin !== undefined && object.offerCoin !== null) {
            message.offerCoin = object.offerCoin;
        }
        else {
            message.offerCoin = "";
        }
        if (object.askDenom !== undefined && object.askDenom !== null) {
            message.askDenom = object.askDenom;
        }
        else {
            message.askDenom = "";
        }
        return message;
    },
};
const baseQuerySwapResponse = {};
exports.QuerySwapResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.returnCoin !== undefined) {
            coin_1.Coin.encode(message.returnCoin, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = Object.assign({}, baseQuerySwapResponse);
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.returnCoin = coin_1.Coin.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const message = Object.assign({}, baseQuerySwapResponse);
        if (object.returnCoin !== undefined && object.returnCoin !== null) {
            message.returnCoin = coin_1.Coin.fromJSON(object.returnCoin);
        }
        else {
            message.returnCoin = undefined;
        }
        return message;
    },
    toJSON(message) {
        const obj = {};
        message.returnCoin !== undefined &&
            (obj.returnCoin = message.returnCoin ? coin_1.Coin.toJSON(message.returnCoin) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = Object.assign({}, baseQuerySwapResponse);
        if (object.returnCoin !== undefined && object.returnCoin !== null) {
            message.returnCoin = coin_1.Coin.fromPartial(object.returnCoin);
        }
        else {
            message.returnCoin = undefined;
        }
        return message;
    },
};
const baseQueryTerraPoolDeltaRequest = {};
exports.QueryTerraPoolDeltaRequest = {
    encode(_, writer = minimal_1.default.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = Object.assign({}, baseQueryTerraPoolDeltaRequest);
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(_) {
        const message = Object.assign({}, baseQueryTerraPoolDeltaRequest);
        return message;
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    fromPartial(_) {
        const message = Object.assign({}, baseQueryTerraPoolDeltaRequest);
        return message;
    },
};
const baseQueryTerraPoolDeltaResponse = {};
exports.QueryTerraPoolDeltaResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.terraPoolDelta.length !== 0) {
            writer.uint32(10).bytes(message.terraPoolDelta);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = Object.assign({}, baseQueryTerraPoolDeltaResponse);
        message.terraPoolDelta = new Uint8Array();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.terraPoolDelta = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const message = Object.assign({}, baseQueryTerraPoolDeltaResponse);
        message.terraPoolDelta = new Uint8Array();
        if (object.terraPoolDelta !== undefined && object.terraPoolDelta !== null) {
            message.terraPoolDelta = bytesFromBase64(object.terraPoolDelta);
        }
        return message;
    },
    toJSON(message) {
        const obj = {};
        message.terraPoolDelta !== undefined &&
            (obj.terraPoolDelta = base64FromBytes(message.terraPoolDelta !== undefined ? message.terraPoolDelta : new Uint8Array()));
        return obj;
    },
    fromPartial(object) {
        const message = Object.assign({}, baseQueryTerraPoolDeltaResponse);
        if (object.terraPoolDelta !== undefined && object.terraPoolDelta !== null) {
            message.terraPoolDelta = object.terraPoolDelta;
        }
        else {
            message.terraPoolDelta = new Uint8Array();
        }
        return message;
    },
};
const baseQueryParamsRequest = {};
exports.QueryParamsRequest = {
    encode(_, writer = minimal_1.default.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = Object.assign({}, baseQueryParamsRequest);
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(_) {
        const message = Object.assign({}, baseQueryParamsRequest);
        return message;
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    fromPartial(_) {
        const message = Object.assign({}, baseQueryParamsRequest);
        return message;
    },
};
const baseQueryParamsResponse = {};
exports.QueryParamsResponse = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.params !== undefined) {
            market_1.Params.encode(message.params, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = Object.assign({}, baseQueryParamsResponse);
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.params = market_1.Params.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const message = Object.assign({}, baseQueryParamsResponse);
        if (object.params !== undefined && object.params !== null) {
            message.params = market_1.Params.fromJSON(object.params);
        }
        else {
            message.params = undefined;
        }
        return message;
    },
    toJSON(message) {
        const obj = {};
        message.params !== undefined && (obj.params = message.params ? market_1.Params.toJSON(message.params) : undefined);
        return obj;
    },
    fromPartial(object) {
        const message = Object.assign({}, baseQueryParamsResponse);
        if (object.params !== undefined && object.params !== null) {
            message.params = market_1.Params.fromPartial(object.params);
        }
        else {
            message.params = undefined;
        }
        return message;
    },
};
class QueryClientImpl {
    constructor(rpc) {
        this.rpc = rpc;
        this.Swap = this.Swap.bind(this);
        this.TerraPoolDelta = this.TerraPoolDelta.bind(this);
        this.Params = this.Params.bind(this);
    }
    Swap(request) {
        const data = exports.QuerySwapRequest.encode(request).finish();
        const promise = this.rpc.request("terra.market.v1beta1.Query", "Swap", data);
        return promise.then((data) => exports.QuerySwapResponse.decode(new minimal_1.default.Reader(data)));
    }
    TerraPoolDelta(request) {
        const data = exports.QueryTerraPoolDeltaRequest.encode(request).finish();
        const promise = this.rpc.request("terra.market.v1beta1.Query", "TerraPoolDelta", data);
        return promise.then((data) => exports.QueryTerraPoolDeltaResponse.decode(new minimal_1.default.Reader(data)));
    }
    Params(request) {
        const data = exports.QueryParamsRequest.encode(request).finish();
        const promise = this.rpc.request("terra.market.v1beta1.Query", "Params", data);
        return promise.then((data) => exports.QueryParamsResponse.decode(new minimal_1.default.Reader(data)));
    }
}
exports.QueryClientImpl = QueryClientImpl;
var globalThis = (() => {
    if (typeof globalThis !== "undefined")
        return globalThis;
    if (typeof self !== "undefined")
        return self;
    if (typeof window !== "undefined")
        return window;
    if (typeof global !== "undefined")
        return global;
    throw "Unable to locate global object";
})();
const atob = globalThis.atob || ((b64) => globalThis.Buffer.from(b64, "base64").toString("binary"));
function bytesFromBase64(b64) {
    const bin = atob(b64);
    const arr = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; ++i) {
        arr[i] = bin.charCodeAt(i);
    }
    return arr;
}
const btoa = globalThis.btoa || ((bin) => globalThis.Buffer.from(bin, "binary").toString("base64"));
function base64FromBytes(arr) {
    const bin = [];
    for (const byte of arr) {
        bin.push(String.fromCharCode(byte));
    }
    return btoa(bin.join(""));
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
