// @eth-optimism/contracts/dist/types/ICanonicalTransactionChain.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import {
  ethers,
  EventFilter,
  Signer,
  BigNumber,
  BigNumberish,
  PopulatedTransaction,
  BaseContract,
  ContractTransaction,
  Overrides,
  CallOverrides,
} from "ethers";
import { BytesLike } from "@ethersproject/bytes";
import { Listener, Provider } from "@ethersproject/providers";
import { FunctionFragment, EventFragment, Result } from "@ethersproject/abi";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "./common";

export type QueueElementStruct = {
  transactionHash: BytesLike;
  timestamp: BigNumberish;
  blockNumber: BigNumberish;
};

export type QueueElementStructOutput = [string, number, number] & {
  transactionHash: string;
  timestamp: number;
  blockNumber: number;
};

export interface ICanonicalTransactionChainInterface
  extends ethers.utils.Interface {
  functions: {
    "appendSequencerBatch()": FunctionFragment;
    "batches()": FunctionFragment;
    "enqueue(address,uint256,bytes)": FunctionFragment;
    "getLastBlockNumber()": FunctionFragment;
    "getLastTimestamp()": FunctionFragment;
    "getNextQueueIndex()": FunctionFragment;
    "getNumPendingQueueElements()": FunctionFragment;
    "getQueueElement(uint256)": FunctionFragment;
    "getQueueLength()": FunctionFragment;
    "getTotalBatches()": FunctionFragment;
    "getTotalElements()": FunctionFragment;
    "setGasParams(uint256,uint256)": FunctionFragment;
  };

  encodeFunctionData(
    functionFragment: "appendSequencerBatch",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "batches", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "enqueue",
    values: [string, BigNumberish, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "getLastBlockNumber",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "getLastTimestamp",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "getNextQueueIndex",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "getNumPendingQueueElements",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "getQueueElement",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "getQueueLength",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "getTotalBatches",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "getTotalElements",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "setGasParams",
    values: [BigNumberish, BigNumberish]
  ): string;

  decodeFunctionResult(
    functionFragment: "appendSequencerBatch",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "batches", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "enqueue", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "getLastBlockNumber",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getLastTimestamp",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getNextQueueIndex",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getNumPendingQueueElements",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getQueueElement",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getQueueLength",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getTotalBatches",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getTotalElements",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setGasParams",
    data: BytesLike
  ): Result;

  events: {
    "L2GasParamsUpdated(uint256,uint256,uint256)": EventFragment;
    "QueueBatchAppended(uint256,uint256,uint256)": EventFragment;
    "SequencerBatchAppended(uint256,uint256,uint256)": EventFragment;
    "TransactionBatchAppended(uint256,bytes32,uint256,uint256,bytes)": EventFragment;
    "TransactionEnqueued(address,address,uint256,bytes,uint256,uint256)": EventFragment;
  };

  getEvent(nameOrSignatureOrTopic: "L2GasParamsUpdated"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "QueueBatchAppended"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "SequencerBatchAppended"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "TransactionBatchAppended"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "TransactionEnqueued"): EventFragment;
}

export type L2GasParamsUpdatedEvent = TypedEvent<
  [BigNumber, BigNumber, BigNumber],
  {
    l2GasDiscountDivisor: BigNumber;
    enqueueGasCost: BigNumber;
    enqueueL2GasPrepaid: BigNumber;
  }
>;

export type L2GasParamsUpdatedEventFilter =
  TypedEventFilter<L2GasParamsUpdatedEvent>;

export type QueueBatchAppendedEvent = TypedEvent<
  [BigNumber, BigNumber, BigNumber],
  {
    _startingQueueIndex: BigNumber;
    _numQueueElements: BigNumber;
    _totalElements: BigNumber;
  }
>;

export type QueueBatchAppendedEventFilter =
  TypedEventFilter<QueueBatchAppendedEvent>;

export type SequencerBatchAppendedEvent = TypedEvent<
  [BigNumber, BigNumber, BigNumber],
  {
    _startingQueueIndex: BigNumber;
    _numQueueElements: BigNumber;
    _totalElements: BigNumber;
  }
>;

export type SequencerBatchAppendedEventFilter =
  TypedEventFilter<SequencerBatchAppendedEvent>;

export type TransactionBatchAppendedEvent = TypedEvent<
  [BigNumber, string, BigNumber, BigNumber, string],
  {
    _batchIndex: BigNumber;
    _batchRoot: string;
    _batchSize: BigNumber;
    _prevTotalElements: BigNumber;
    _extraData: string;
  }
>;

export type TransactionBatchAppendedEventFilter =
  TypedEventFilter<TransactionBatchAppendedEvent>;

export type TransactionEnqueuedEvent = TypedEvent<
  [string, string, BigNumber, string, BigNumber, BigNumber],
  {
    _l1TxOrigin: string;
    _target: string;
    _gasLimit: BigNumber;
    _data: string;
    _queueIndex: BigNumber;
    _timestamp: BigNumber;
  }
>;

export type TransactionEnqueuedEventFilter =
  TypedEventFilter<TransactionEnqueuedEvent>;

export interface ICanonicalTransactionChain extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: ICanonicalTransactionChainInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    appendSequencerBatch(
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;

    batches(overrides?: CallOverrides): Promise<[string]>;

    enqueue(
      _target: string,
      _gasLimit: BigNumberish,
      _data: BytesLike,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;

    getLastBlockNumber(overrides?: CallOverrides): Promise<[number]>;

    getLastTimestamp(overrides?: CallOverrides): Promise<[number]>;

    getNextQueueIndex(overrides?: CallOverrides): Promise<[number]>;

    getNumPendingQueueElements(overrides?: CallOverrides): Promise<[number]>;

    getQueueElement(
      _index: BigNumberish,
      overrides?: CallOverrides
    ): Promise<
      [QueueElementStructOutput] & { _element: QueueElementStructOutput }
    >;

    getQueueLength(overrides?: CallOverrides): Promise<[number]>;

    getTotalBatches(
      overrides?: CallOverrides
    ): Promise<[BigNumber] & { _totalBatches: BigNumber }>;

    getTotalElements(
      overrides?: CallOverrides
    ): Promise<[BigNumber] & { _totalElements: BigNumber }>;

    setGasParams(
      _l2GasDiscountDivisor: BigNumberish,
      _enqueueGasCost: BigNumberish,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;
  };

  appendSequencerBatch(
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  batches(overrides?: CallOverrides): Promise<string>;

  enqueue(
    _target: string,
    _gasLimit: BigNumberish,
    _data: BytesLike,
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  getLastBlockNumber(overrides?: CallOverrides): Promise<number>;

  getLastTimestamp(overrides?: CallOverrides): Promise<number>;

  getNextQueueIndex(overrides?: CallOverrides): Promise<number>;

  getNumPendingQueueElements(overrides?: CallOverrides): Promise<number>;

  getQueueElement(
    _index: BigNumberish,
    overrides?: CallOverrides
  ): Promise<QueueElementStructOutput>;

  getQueueLength(overrides?: CallOverrides): Promise<number>;

  getTotalBatches(overrides?: CallOverrides): Promise<BigNumber>;

  getTotalElements(overrides?: CallOverrides): Promise<BigNumber>;

  setGasParams(
    _l2GasDiscountDivisor: BigNumberish,
    _enqueueGasCost: BigNumberish,
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    appendSequencerBatch(overrides?: CallOverrides): Promise<void>;

    batches(overrides?: CallOverrides): Promise<string>;

    enqueue(
      _target: string,
      _gasLimit: BigNumberish,
      _data: BytesLike,
      overrides?: CallOverrides
    ): Promise<void>;

    getLastBlockNumber(overrides?: CallOverrides): Promise<number>;

    getLastTimestamp(overrides?: CallOverrides): Promise<number>;

    getNextQueueIndex(overrides?: CallOverrides): Promise<number>;

    getNumPendingQueueElements(overrides?: CallOverrides): Promise<number>;

    getQueueElement(
      _index: BigNumberish,
      overrides?: CallOverrides
    ): Promise<QueueElementStructOutput>;

    getQueueLength(overrides?: CallOverrides): Promise<number>;

    getTotalBatches(overrides?: CallOverrides): Promise<BigNumber>;

    getTotalElements(overrides?: CallOverrides): Promise<BigNumber>;

    setGasParams(
      _l2GasDiscountDivisor: BigNumberish,
      _enqueueGasCost: BigNumberish,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {
    "L2GasParamsUpdated(uint256,uint256,uint256)"(
      l2GasDiscountDivisor?: null,
      enqueueGasCost?: null,
      enqueueL2GasPrepaid?: null
    ): L2GasParamsUpdatedEventFilter;
    L2GasParamsUpdated(
      l2GasDiscountDivisor?: null,
      enqueueGasCost?: null,
      enqueueL2GasPrepaid?: null
    ): L2GasParamsUpdatedEventFilter;

    "QueueBatchAppended(uint256,uint256,uint256)"(
      _startingQueueIndex?: null,
      _numQueueElements?: null,
      _totalElements?: null
    ): QueueBatchAppendedEventFilter;
    QueueBatchAppended(
      _startingQueueIndex?: null,
      _numQueueElements?: null,
      _totalElements?: null
    ): QueueBatchAppendedEventFilter;

    "SequencerBatchAppended(uint256,uint256,uint256)"(
      _startingQueueIndex?: null,
      _numQueueElements?: null,
      _totalElements?: null
    ): SequencerBatchAppendedEventFilter;
    SequencerBatchAppended(
      _startingQueueIndex?: null,
      _numQueueElements?: null,
      _totalElements?: null
    ): SequencerBatchAppendedEventFilter;

    "TransactionBatchAppended(uint256,bytes32,uint256,uint256,bytes)"(
      _batchIndex?: BigNumberish | null,
      _batchRoot?: null,
      _batchSize?: null,
      _prevTotalElements?: null,
      _extraData?: null
    ): TransactionBatchAppendedEventFilter;
    TransactionBatchAppended(
      _batchIndex?: BigNumberish | null,
      _batchRoot?: null,
      _batchSize?: null,
      _prevTotalElements?: null,
      _extraData?: null
    ): TransactionBatchAppendedEventFilter;

    "TransactionEnqueued(address,address,uint256,bytes,uint256,uint256)"(
      _l1TxOrigin?: string | null,
      _target?: string | null,
      _gasLimit?: null,
      _data?: null,
      _queueIndex?: BigNumberish | null,
      _timestamp?: null
    ): TransactionEnqueuedEventFilter;
    TransactionEnqueued(
      _l1TxOrigin?: string | null,
      _target?: string | null,
      _gasLimit?: null,
      _data?: null,
      _queueIndex?: BigNumberish | null,
      _timestamp?: null
    ): TransactionEnqueuedEventFilter;
  };

  estimateGas: {
    appendSequencerBatch(
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;

    batches(overrides?: CallOverrides): Promise<BigNumber>;

    enqueue(
      _target: string,
      _gasLimit: BigNumberish,
      _data: BytesLike,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;

    getLastBlockNumber(overrides?: CallOverrides): Promise<BigNumber>;

    getLastTimestamp(overrides?: CallOverrides): Promise<BigNumber>;

    getNextQueueIndex(overrides?: CallOverrides): Promise<BigNumber>;

    getNumPendingQueueElements(overrides?: CallOverrides): Promise<BigNumber>;

    getQueueElement(
      _index: BigNumberish,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    getQueueLength(overrides?: CallOverrides): Promise<BigNumber>;

    getTotalBatches(overrides?: CallOverrides): Promise<BigNumber>;

    getTotalElements(overrides?: CallOverrides): Promise<BigNumber>;

    setGasParams(
      _l2GasDiscountDivisor: BigNumberish,
      _enqueueGasCost: BigNumberish,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    appendSequencerBatch(
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;

    batches(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    enqueue(
      _target: string,
      _gasLimit: BigNumberish,
      _data: BytesLike,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;

    getLastBlockNumber(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    getLastTimestamp(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    getNextQueueIndex(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    getNumPendingQueueElements(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    getQueueElement(
      _index: BigNumberish,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    getQueueLength(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    getTotalBatches(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    getTotalElements(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    setGasParams(
      _l2GasDiscountDivisor: BigNumberish,
      _enqueueGasCost: BigNumberish,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;
  };
}
