// @eth-optimism/contracts/dist/types/L2StandardTokenFactory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import {
  ethers,
  EventFilter,
  Signer,
  BigNumber,
  BigNumberish,
  PopulatedTransaction,
  BaseContract,
  ContractTransaction,
  Overrides,
  CallOverrides,
} from "ethers";
import { BytesLike } from "@ethersproject/bytes";
import { Listener, Provider } from "@ethersproject/providers";
import { FunctionFragment, EventFragment, Result } from "@ethersproject/abi";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "./common";

export interface L2StandardTokenFactoryInterface
  extends ethers.utils.Interface {
  functions: {
    "createStandardL2Token(address,string,string)": FunctionFragment;
  };

  encodeFunctionData(
    functionFragment: "createStandardL2Token",
    values: [string, string, string]
  ): string;

  decodeFunctionResult(
    functionFragment: "createStandardL2Token",
    data: BytesLike
  ): Result;

  events: {
    "StandardL2TokenCreated(address,address)": EventFragment;
  };

  getEvent(nameOrSignatureOrTopic: "StandardL2TokenCreated"): EventFragment;
}

export type StandardL2TokenCreatedEvent = TypedEvent<
  [string, string],
  { _l1Token: string; _l2Token: string }
>;

export type StandardL2TokenCreatedEventFilter =
  TypedEventFilter<StandardL2TokenCreatedEvent>;

export interface L2StandardTokenFactory extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: L2StandardTokenFactoryInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    createStandardL2Token(
      _l1Token: string,
      _name: string,
      _symbol: string,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;
  };

  createStandardL2Token(
    _l1Token: string,
    _name: string,
    _symbol: string,
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    createStandardL2Token(
      _l1Token: string,
      _name: string,
      _symbol: string,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {
    "StandardL2TokenCreated(address,address)"(
      _l1Token?: string | null,
      _l2Token?: string | null
    ): StandardL2TokenCreatedEventFilter;
    StandardL2TokenCreated(
      _l1Token?: string | null,
      _l2Token?: string | null
    ): StandardL2TokenCreatedEventFilter;
  };

  estimateGas: {
    createStandardL2Token(
      _l1Token: string,
      _name: string,
      _symbol: string,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    createStandardL2Token(
      _l1Token: string,
      _name: string,
      _symbol: string,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;
  };
}
