// @eth-optimism/contracts/dist/types/TestLibBytes32Utils.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import {
  ethers,
  EventFilter,
  Signer,
  BigNumber,
  BigNumberish,
  PopulatedTransaction,
  BaseContract,
  ContractTransaction,
  CallOverrides,
} from "ethers";
import { BytesLike } from "@ethersproject/bytes";
import { Listener, Provider } from "@ethersproject/providers";
import { FunctionFragment, EventFragment, Result } from "@ethersproject/abi";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "./common";

export interface TestLibBytes32UtilsInterface extends ethers.utils.Interface {
  functions: {
    "fromAddress(address)": FunctionFragment;
    "fromBool(bool)": FunctionFragment;
    "toAddress(bytes32)": FunctionFragment;
    "toBool(bytes32)": FunctionFragment;
  };

  encodeFunctionData(functionFragment: "fromAddress", values: [string]): string;
  encodeFunctionData(functionFragment: "fromBool", values: [boolean]): string;
  encodeFunctionData(
    functionFragment: "toAddress",
    values: [BytesLike]
  ): string;
  encodeFunctionData(functionFragment: "toBool", values: [BytesLike]): string;

  decodeFunctionResult(
    functionFragment: "fromAddress",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "fromBool", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "toAddress", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "toBool", data: BytesLike): Result;

  events: {};
}

export interface TestLibBytes32Utils extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: TestLibBytes32UtilsInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    fromAddress(
      _in: string,
      overrides?: CallOverrides
    ): Promise<[string] & { _out: string }>;

    fromBool(
      _in: boolean,
      overrides?: CallOverrides
    ): Promise<[string] & { _out: string }>;

    toAddress(
      _in: BytesLike,
      overrides?: CallOverrides
    ): Promise<[string] & { _out: string }>;

    toBool(
      _in: BytesLike,
      overrides?: CallOverrides
    ): Promise<[boolean] & { _out: boolean }>;
  };

  fromAddress(_in: string, overrides?: CallOverrides): Promise<string>;

  fromBool(_in: boolean, overrides?: CallOverrides): Promise<string>;

  toAddress(_in: BytesLike, overrides?: CallOverrides): Promise<string>;

  toBool(_in: BytesLike, overrides?: CallOverrides): Promise<boolean>;

  callStatic: {
    fromAddress(_in: string, overrides?: CallOverrides): Promise<string>;

    fromBool(_in: boolean, overrides?: CallOverrides): Promise<string>;

    toAddress(_in: BytesLike, overrides?: CallOverrides): Promise<string>;

    toBool(_in: BytesLike, overrides?: CallOverrides): Promise<boolean>;
  };

  filters: {};

  estimateGas: {
    fromAddress(_in: string, overrides?: CallOverrides): Promise<BigNumber>;

    fromBool(_in: boolean, overrides?: CallOverrides): Promise<BigNumber>;

    toAddress(_in: BytesLike, overrides?: CallOverrides): Promise<BigNumber>;

    toBool(_in: BytesLike, overrides?: CallOverrides): Promise<BigNumber>;
  };

  populateTransaction: {
    fromAddress(
      _in: string,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    fromBool(
      _in: boolean,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    toAddress(
      _in: BytesLike,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    toBool(
      _in: BytesLike,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;
  };
}
