// @eth-optimism/contracts/dist/types/TestLibBuffer.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import {
  ethers,
  EventFilter,
  Signer,
  BigNumber,
  BigNumberish,
  PopulatedTransaction,
  BaseContract,
  ContractTransaction,
  Overrides,
  CallOverrides,
} from "ethers";
import { BytesLike } from "@ethersproject/bytes";
import { Listener, Provider } from "@ethersproject/providers";
import { FunctionFragment, EventFragment, Result } from "@ethersproject/abi";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "./common";

export type BufferContextStruct = {
  length: BigNumberish;
  extraData: BytesLike;
};

export type BufferContextStructOutput = [number, string] & {
  length: number;
  extraData: string;
};

export interface TestLibBufferInterface extends ethers.utils.Interface {
  functions: {
    "deleteElementsAfterInclusive(uint40,bytes27)": FunctionFragment;
    "get(uint256)": FunctionFragment;
    "getContext()": FunctionFragment;
    "getExtraData()": FunctionFragment;
    "getLength()": FunctionFragment;
    "push(bytes32,bytes27)": FunctionFragment;
    "setContext(uint40,bytes27)": FunctionFragment;
    "setExtraData(bytes27)": FunctionFragment;
  };

  encodeFunctionData(
    functionFragment: "deleteElementsAfterInclusive",
    values: [BigNumberish, BytesLike]
  ): string;
  encodeFunctionData(functionFragment: "get", values: [BigNumberish]): string;
  encodeFunctionData(
    functionFragment: "getContext",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "getExtraData",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "getLength", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "push",
    values: [BytesLike, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "setContext",
    values: [BigNumberish, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "setExtraData",
    values: [BytesLike]
  ): string;

  decodeFunctionResult(
    functionFragment: "deleteElementsAfterInclusive",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "get", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "getContext", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "getExtraData",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "getLength", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "push", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "setContext", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "setExtraData",
    data: BytesLike
  ): Result;

  events: {};
}

export interface TestLibBuffer extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: TestLibBufferInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    "deleteElementsAfterInclusive(uint40,bytes27)"(
      _index: BigNumberish,
      _extraData: BytesLike,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;

    "deleteElementsAfterInclusive(uint40)"(
      _index: BigNumberish,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;

    get(_index: BigNumberish, overrides?: CallOverrides): Promise<[string]>;

    getContext(overrides?: CallOverrides): Promise<[BufferContextStructOutput]>;

    getExtraData(overrides?: CallOverrides): Promise<[string]>;

    getLength(overrides?: CallOverrides): Promise<[number]>;

    "push(bytes32,bytes27)"(
      _value: BytesLike,
      _extraData: BytesLike,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;

    "push(bytes32)"(
      _value: BytesLike,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;

    setContext(
      _index: BigNumberish,
      _extraData: BytesLike,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;

    setExtraData(
      _extraData: BytesLike,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;
  };

  "deleteElementsAfterInclusive(uint40,bytes27)"(
    _index: BigNumberish,
    _extraData: BytesLike,
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  "deleteElementsAfterInclusive(uint40)"(
    _index: BigNumberish,
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  get(_index: BigNumberish, overrides?: CallOverrides): Promise<string>;

  getContext(overrides?: CallOverrides): Promise<BufferContextStructOutput>;

  getExtraData(overrides?: CallOverrides): Promise<string>;

  getLength(overrides?: CallOverrides): Promise<number>;

  "push(bytes32,bytes27)"(
    _value: BytesLike,
    _extraData: BytesLike,
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  "push(bytes32)"(
    _value: BytesLike,
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  setContext(
    _index: BigNumberish,
    _extraData: BytesLike,
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  setExtraData(
    _extraData: BytesLike,
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    "deleteElementsAfterInclusive(uint40,bytes27)"(
      _index: BigNumberish,
      _extraData: BytesLike,
      overrides?: CallOverrides
    ): Promise<void>;

    "deleteElementsAfterInclusive(uint40)"(
      _index: BigNumberish,
      overrides?: CallOverrides
    ): Promise<void>;

    get(_index: BigNumberish, overrides?: CallOverrides): Promise<string>;

    getContext(overrides?: CallOverrides): Promise<BufferContextStructOutput>;

    getExtraData(overrides?: CallOverrides): Promise<string>;

    getLength(overrides?: CallOverrides): Promise<number>;

    "push(bytes32,bytes27)"(
      _value: BytesLike,
      _extraData: BytesLike,
      overrides?: CallOverrides
    ): Promise<void>;

    "push(bytes32)"(
      _value: BytesLike,
      overrides?: CallOverrides
    ): Promise<void>;

    setContext(
      _index: BigNumberish,
      _extraData: BytesLike,
      overrides?: CallOverrides
    ): Promise<void>;

    setExtraData(
      _extraData: BytesLike,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {};

  estimateGas: {
    "deleteElementsAfterInclusive(uint40,bytes27)"(
      _index: BigNumberish,
      _extraData: BytesLike,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;

    "deleteElementsAfterInclusive(uint40)"(
      _index: BigNumberish,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;

    get(_index: BigNumberish, overrides?: CallOverrides): Promise<BigNumber>;

    getContext(overrides?: CallOverrides): Promise<BigNumber>;

    getExtraData(overrides?: CallOverrides): Promise<BigNumber>;

    getLength(overrides?: CallOverrides): Promise<BigNumber>;

    "push(bytes32,bytes27)"(
      _value: BytesLike,
      _extraData: BytesLike,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;

    "push(bytes32)"(
      _value: BytesLike,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;

    setContext(
      _index: BigNumberish,
      _extraData: BytesLike,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;

    setExtraData(
      _extraData: BytesLike,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    "deleteElementsAfterInclusive(uint40,bytes27)"(
      _index: BigNumberish,
      _extraData: BytesLike,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;

    "deleteElementsAfterInclusive(uint40)"(
      _index: BigNumberish,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;

    get(
      _index: BigNumberish,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    getContext(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    getExtraData(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    getLength(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    "push(bytes32,bytes27)"(
      _value: BytesLike,
      _extraData: BytesLike,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;

    "push(bytes32)"(
      _value: BytesLike,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;

    setContext(
      _index: BigNumberish,
      _extraData: BytesLike,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;

    setExtraData(
      _extraData: BytesLike,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;
  };
}
