// @chainify/solana/dist/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolanaWalletProvider = exports.SolanaUtils = exports.SolanaTypes = exports.SolanaNftProvider = exports.SolanaNetworks = exports.SolanaChainProvider = void 0;
var SolanaChainProvider_1 = require("./chain/SolanaChainProvider");
Object.defineProperty(exports, "SolanaChainProvider", { enumerable: true, get: function () { return SolanaChainProvider_1.SolanaChainProvider; } });
var networks_1 = require("./networks");
Object.defineProperty(exports, "SolanaNetworks", { enumerable: true, get: function () { return networks_1.SolanaNetworks; } });
var SolanaNftProvider_1 = require("./nft/SolanaNftProvider");
Object.defineProperty(exports, "SolanaNftProvider", { enumerable: true, get: function () { return SolanaNftProvider_1.SolanaNftProvider; } });
exports.SolanaTypes = __importStar(require("./types"));
exports.SolanaUtils = __importStar(require("./utils"));
var SolanaWalletProvider_1 = require("./wallet/SolanaWalletProvider");
Object.defineProperty(exports, "SolanaWalletProvider", { enumerable: true, get: function () { return SolanaWalletProvider_1.SolanaWalletProvider; } });
