// @chainify/utils/dist/lib/math.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.gte = exports.lt = exports.lte = exports.eq = exports.div = exports.mul = exports.sub = exports.add = void 0;
const types_1 = require("@chainify/types");
function add(a, b) {
    return new types_1.BigNumber(a.toString()).plus(b.toString());
}
exports.add = add;
function sub(a, b) {
    return new types_1.BigNumber(a.toString()).minus(b.toString());
}
exports.sub = sub;
function mul(a, b) {
    return new types_1.BigNumber(a.toString()).multipliedBy(b.toString());
}
exports.mul = mul;
function div(a, b) {
    return new types_1.BigNumber(a.toString()).dividedBy(b.toString());
}
exports.div = div;
function eq(a, b) {
    return new types_1.BigNumber(a.toString()).eq(b.toString());
}
exports.eq = eq;
function lte(a, b) {
    return new types_1.BigNumber(a.toString()).lte(b.toString());
}
exports.lte = lte;
function lt(a, b) {
    return new types_1.BigNumber(a.toString()).lt(b.toString());
}
exports.lt = lt;
function gte(a, b) {
    return new types_1.BigNumber(a.toString()).gte(b.toString());
}
exports.gte = gte;
