// @chainify/evm/dist/lib/fee/EIP1559FeeApiProvider/polygon.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PolygonFeeParser = void 0;
const errors_1 = require("@chainify/errors");
const utils_1 = require("@chainify/utils");
class PolygonFeeParser {
    parse(response) {
        if (response.status === '1' && response.message === 'OK') {
            const suggestedBaseFeePerGas = Number(response.result.suggestBaseFee);
            return {
                slow: {
                    fee: {
                        maxFeePerGas: Number(response.result.SafeGasPrice),
                        maxPriorityFeePerGas: utils_1.Math.sub(response.result.SafeGasPrice, suggestedBaseFeePerGas).toNumber(),
                        suggestedBaseFeePerGas,
                    },
                },
                average: {
                    fee: {
                        maxFeePerGas: Number(response.result.ProposeGasPrice),
                        maxPriorityFeePerGas: utils_1.Math.sub(response.result.ProposeGasPrice, suggestedBaseFeePerGas).toNumber(),
                        suggestedBaseFeePerGas,
                    },
                },
                fast: {
                    fee: {
                        maxFeePerGas: Number(response.result.FastGasPrice),
                        maxPriorityFeePerGas: utils_1.Math.sub(response.result.FastGasPrice, suggestedBaseFeePerGas).toNumber(),
                        suggestedBaseFeePerGas,
                    },
                },
            };
        }
        else {
            throw new errors_1.NodeError('Could not fetch Polygon fee data', response);
        }
    }
}
exports.PolygonFeeParser = PolygonFeeParser;
