// @sinatdt/configs/dist/teleswap/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.nodeInfo = exports.ABI = exports.tokenInfo = exports.contractsInfo = exports.requestTypes = exports.bitcoinAddressTypes = void 0;
var bitcoin_address_types_1 = require("./bitcoin-address-types");
Object.defineProperty(exports, "bitcoinAddressTypes", { enumerable: true, get: function () { return __importDefault(bitcoin_address_types_1).default; } });
var request_types_1 = require("./request-types");
Object.defineProperty(exports, "requestTypes", { enumerable: true, get: function () { return __importDefault(request_types_1).default; } });
exports.contractsInfo = __importStar(require("./contracts-info"));
exports.tokenInfo = __importStar(require("./tokens-info"));
exports.ABI = __importStar(require("./abis"));
exports.nodeInfo = __importStar(require("./node-addresses"));
